// hqchart/src/utils/urlObj.js
import axios from 'axios'
// import tools from "../services/tools"

var instance = axios.create({
  timeout: 200000,
  headers: {
    'Content-Type': 'application/json'
    // "Authorization": 'Basic ' + tools.getCookie('HXtoken')
  },
  // transformResponse: [function (res) {
  //     // 在此转码数据
  //     return res;
  // }],
});
instance.interceptors.response.use(function (response) {
  // 对响应数据做点什么
  return response.data;
}, function (error) {
  // 对响应错误做点什么
  console.log("对响应错误做点什么",error);
  return Promise.reject(error);
});
instance.interceptors.request.use(function (config) {
  // config.headers = {
  //   "Authorization": 'Basic ' + tools.getCookie('HXtoken')
  // }
  return config
}, function (error) {
  // 对req错误做点什么
  console.log("对req错误做点什么",error);
  return Promise.reject(error);
});
var post = function (apiurl, data, fn,err) {
  return instance.post(apiurl, data).then(res => {
    fn(res)
      return res;
  }).catch(e => {
      err && err(e);
  });
}
var get = function (url, data, fn) {
  return instance.get(url, data).then(res => {
    // console.log(`fn:${fn}`);
      fn(res)
      return res;
  });
}

const assemblUrl = function(url){
  return '.' + url;
}

const assemblApi = function(url){
  // deBug: "https://hexun-api.zealink.com",
  var apiDomain = {
    release: "",
    deBug: "https://opensource.zealink.com",
    isRelease : false
  };

  var api = apiDomain.isRelease ? apiDomain.release : apiDomain.deBug;

  if (url == undefined) {
    return api;
  }
  return api + url;
}


const urlObj = {
  //方法
  post: post,
  get: get,
  assemblApi:assemblApi,
  assemblUrl:assemblUrl,
  
  apiStockDetail:assemblApi("/API/StockDetail"),
  apiStockPriceList: assemblApi('/API/StockPriceList')
}

export default urlObj;
