// hqchart/src/jscommon/umychart.wechat/lighter.news.wechat.js

function  QuickNews() {

}

//获取新闻列表
QuickNews.GetNewsList = function (newsType) {
  switch (newsType) {
    case NEWS_TYPE.STOCK_NEWS:
      return new GetNews();
      break;
  }
}
QuickNews.GetSeyyingData = function (symbol,newsType){
  switch (newsType) {
    case NEWS_TYPE.SETTING:
      return new GetTactics(symbol);
      break;
  }
}

//类型
var NEWS_TYPE =
  {
    STOCK_NEWS: 1,             //快讯
    SETTING:2                  //设置
  }



//快讯基类
function News(){

  this.Data = new Array();  //快讯数据 
  this.Callback;
  this.Count;                 //一共的快讯个数
  this.CacheCount;            //已下载的快讯个数
  this.PageSize = 10;           //每页的快讯个数
  this.PageIndex = 1;
  this.ApiUrl;                //数据请求地址

  this.startDate;             //开始时间
  this.endDate;               //结束时间

  //是否是最后一页
  this.IsEndPage = function () {
    if (this.Count == null || this.Count <= 0) return true;

    return this.CacheCount >= this.Count;

  }

  //获取第1页
  this.GetFirstPage = function () {

  }

  //下拉获取下一页
  this.GetNextPage = function () {

  }
}

//策略应对基类
function Tactics(symbol){
  this.Symbol = symbol;         //股票代码
  this.Name;                    //股票名称
  this.Callback;                //回调 function(info,news) info: {Start:请求的新闻起始位置 , End:新闻的结束位置 }
  this.ApiUrl;              //请求地址
  this.Data;      //数据 { Symbol:代码, Name:名称, Title:标题, Content:内容, Date:日期 .....}
  this.Error;     //如果调用失败 把错误信息写这里
  
  this.GetData = function () {

  }
}



//请求快讯数据
function GetNews(){
  this.newMethod = News;   //派生
  this.newMethod();
  delete this.newMethod;

  this.ApiUrl = 'https://opensource.zealink.com/api/LG_NewsFlashList';

  var date = new Date();
  this.endDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();     //获取结束时间
  this.startDate = (date.getFullYear() - 1) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

  //获取第1页
  this.GetFirstPage = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.ApiUrl,
      data:{
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": 10
        },
        "Filter": {
          "QueryDate": {
            "StartDate": this.startDate,
            "EndDate": this.endDate
          }
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        self.RecvData(data);
      },
      fail: function (request) {
        this.RecvError(request);
      }
    })
  },
  this.RecvData = function (recvData)
  {
    var data = recvData.data;
    if (this.Count <= 0)
      this.Count = data.count;                 //一共的新闻个数

    var info = {};
    info.Start = this.Data.length;
    info.End = this.Data.length;

    for (var i in data.list) {
      var item = data.list[i];
      this.Data.push(
        {
          ID: item.newsid,
          Data: item.date,
          Content: item.content
        });

      if (i > 0)++info.End;
    }

    this.InvokeCallback(info);
  },

  this.RecvError = function (reqeust){
    this.Error = "请求失败";
    this.InvokeCallback();
  }, 

    this.InvokeCallback = function (info){
    if (typeof (this.Callback) != 'function') return;

    this.Callback(info,this);
  }

  this.GetNextPage = function () {
    this.PageIndex ++;
    var self = this;

    var start = this.Data.length;
    var end = start + this.PageSize;

    wx.request({
      url: this.ApiUrl,
      data: {
        "PageInfo": {
          "PageIndex": this.PageIndex,
          "PageSize": 10
        },
        "Filter": {
          "QueryDate": {
            "StartDate": this.startDate,
            "EndDate": this.endDate
          }
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        self.RecvData(data);
      },
      fail: function (request) {
        this.RecvError(request);
      }
    })
  }
}

//请求settings数据
function GetTactics(symbol){
  this.newMethod = Tactics;   //派生
  this.newMethod(symbol);
  delete this.newMethod;

  this.ApiUrl = "http://web4.umydata.com/api/StockPolicy";
  this.GetData = function () {
    var self = this;
    //清空数据
    this.Name = null;
    this.Error = null;
    this.Data = null;

    wx.request({
      url: this.ApiUrl,
      data: {
        "symbol": [this.Symbol],
        "index": ["位置研判-顶部区", "位置研判-风险区", "位置研判-高位区", "位置研判-中位区", "位置研判-低位区", "位置研判-底部区", "形态研判-大盘多头", "形态研判-个股多头", "文字提示-情况1","文字提示-情况2","文字提示-情况3","文字提示-情况4","文字提示-情况5","文字提示-情况6","文字提示-情况7", "文字提示-情况8", "文字提示-情况9", "文字提示-情况10", "文字提示-情况11", "文字提示-情况12", "文字提示-情况13", "文字提示-情况14", "文字提示-情况15","文字提示-情况16", "文字提示-情况17", "文字提示-情况18", "文字提示-情况19", "文字提示-情况20", "文字提示-情况21", "文字提示-情况22", "文字提示-情况23", "文字提示-情况24"],
        "field": ["price","increase"],
        "start": 0,
        "end": 60
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function(recvData){
    var data = recvData.data.list[0];

    //保存所需数据
    this.Data = {};
    this.Data.Name = data.name;
    this.Data.Price = data.fieldlist[0].value;
    this.Data.Increase = data.fieldlist[1].value;
    this.Data.Policylist = data.policylist;

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

}

module.exports =
  {
    JSCommonQuickNews:
    {
      QuickNews: QuickNews,
      GetNews: GetNews,
      GetTactics: GetTactics,
      NEWS_TYPE: NEWS_TYPE
    }
  };
