// hqchart/src/jscommon/umychart.wechat/umychart.markethot.js
function Benford() {
  this.symbol = "";
  this.StockHistoryDay = "https://opensource.zealink.com/API/Stock";

}
Benford.prototype.getBenford = function (symbol, callback) {
  let request = new Promise((resolve, reject) => {
    wx.request({
      url: this.StockHistoryDay,
      data: {
        "symbol": [symbol],
        "field": ["name", "symbol", "events.start"]
      },
      method: 'POST',
      success: res => {
        resolve(res);
        this.saveInfo(res, callback)
      },
      fail: error => {
        reject(error);
      }
    })
  })
  return request;
}

Benford.prototype.saveInfo = function (res, callback) {
  if (res.data.count == 0 || res.data.stock.length == 0) return;
  let symbol = res.data.stock[0].symbol;
  let name = res.data.stock[0].name;
  let start = res.data.stock[0].events.start;
  
  this.benford = {
    "symbol": symbol,
    "name": name,
    "start":start
  }
  callback(this.benford);
}

function getBenfordObject() {
  return new Benford();
}

module.exports = {
  financialInfo: {
    getBenfordObject
  }
}
