// hqchart/src/jscommon/umychart.wechat/umychart.simulatetrade.wechat.js
///////////////////////////////////////////////////////////////////////////////////
//  仿真交易接口模块
//
///////////////////////////////////////////////////////////////////////////////////
var g_VTradeResource=
{
  Domain:"https://opensource.zealink.com"         //API域名
}

function SimulateTrade(userid)
{
    this.UserID=userid; //账户id
    
    //创建仿真账户 data={...,Callback:回调}
    this.Register=function(data)
    {
      
    }

    //委托买 buyData={Symbol:股票代码, Name:名称, Price:买价格, Vol:量, Callback:回调}
    this.Buy=function(buyData)
    {
      var Url = g_VTradeResource.Domain + "/API/VTradeBuy";
      wx.request({
        url: Url,
        data:
          {
            "userid": this.UserID, 
            "symbol": buyData.Symbol, 
            "name": buyData.Name, 
            "price": buyData.Price, 
            "vol": buyData.Vol
          },
        method: "POST",
        dataType: "json",
        success: function (data) {
          if (data.data.code != 0) {
            wx.showModal({
              title: '提示',
              content: data.data.message,
            })
            return;
          }
          buyData.Callback(data);
        },
        error: function (request) {
          console.log("error:",request)
        }
      })
    }

    //股票联想
  this.Spell = function (spellData) {
    var Url = g_VTradeResource.Domain + "/API/StockSpell";
    wx.request({
      url: Url,
      data:
        {
          "input": spellData.Input,
          "start": spellData.Start,
          "end": spellData.End
        },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        spellData.Callback(data);
      },
      error: function (request) {
        console.log("error:", request)
      }
    })
  }

    //委托卖 buyData={Symbol:股票代码, Name:名称, Price:卖价格, Vol:量, Callback:回调}
    this.Sell=function(sellData)
    {
      var Url = g_VTradeResource.Domain + "/API/VTradeSell";
      wx.request({
        url: Url,
        data:
          { 
            "userid": this.UserID, 
            "symbol": sellData.Symbol,
            "price": sellData.Price,
            "vol": sellData.Vol
          },
        method: "POST",
        dataType: "json",
        success: function (data) {
          if (data.data.code != 0) {
            wx.showModal({
              title: '提示',
              content: data.data.message,
            })
            return;
          }
          sellData.Callback(data);
        },
        error: function (request) {
          console.log("error:", request)
        }
      })

    }

    //委托撤单 cancelData={BookID:, Callback:回调}
    this.Cancel=function(cancelData)
    {
      var Url = g_VTradeResource.Domain + "/API/VTradeCanel";
      wx.request({
        url: Url,
        data:
          {
            "userid": this.UserID,
            "bookid": cancelData.BookID
          },
        method: "POST",
        dataType: "json",
        success: function (data) {
          if (data.data.code != 0) {
            wx.showModal({
              title: '提示',
              content: data.data.message,
            })
            return;
          }
          cancelData.Callback(data);
        },
        error: function (request) {
          console.log("error:", request)
        }
      })
    }

    //委托查询 queryData={queryType:day0/day1/his0/his1,Type:, Start:起始日期, End:结束日期 Callback:回调}
    this.Query=function(queryData)
    {
      var Url = g_VTradeResource.Domain + "/API/VTradeQuery";
      
      if (queryData.queryType == "day0"){
        var data = 
        {
            "userid": this.UserID, 
            "type": 0
        }
      } else if (queryData.queryType == "day1"){
        var data =
          {
            "userid": this.UserID,
            "type": 1
          }
      } else if (queryData.queryType == "his1") {
        var data =
          {
            "userid": this.UserID,            
            "querydate": { "StartDate": queryData.Start, "EndDate": queryData.End }, 
            "type": 1
          }
      } else if (queryData.queryType == "his0"){
        var data =
          {
            "userid": this.UserID,
            "querydate": { "StartDate": queryData.Start, "EndDate": queryData.End },
            "type": 0
          }
      }

      wx.request({
        url: Url,
        data:data,
        method: "POST",
        dataType: "json",
        success: function (data) {
          if (data.data.code != 0) {
            wx.showModal({
              title: '提示',
              content: data.data.message,
            })
            return;
          }
          queryData.Callback(data);
        },
        error: function (request) {
          console.log("error:", request)
        }
      })
    }

    //账户信息 queryData={Stock:true/false, Callback:回调}
    this.UserInfo=function(userData)
    {
      var Url = g_VTradeResource.Domain + "/API/VTradeUser";
      wx.request({
        url: Url,
        data:
          {
            "userid": this.UserID,
            "stock": userData.Stock
          },
        method: "POST",
        dataType: "json",
        success: function (data) {
          if (data.data.code != 0) {
            wx.showModal({
              title: '提示',
              content: data.data.message,
            })
            return;
          }
          userData.Callback(data);
        },
        error: function (request) {
          console.log("error:", request)
        }
      })
    }
}

SimulateTrade.SetDomain = function (domain, cacheDomain) 
{
    if (domain) g_VTradeResource.Domain = domain;
}

//导出统一使用JSCommon命名空间名
module.exports =
  {
    JSCommonVTrade:
      {
        SimulateTrade: SimulateTrade
      }
  };
