// hqchart/src/jscommon/umychart.wechat/umychart.vedio.wechat.js
var g_VTradeResource =
  {
    Domain: "https://opensource.zealink.com/"         //API域名
  }

//获取视频
function GetVedioList() {
  this.ApiUrl = g_VTradeResource.Domain + 'api/VideoList';
  this.PageIndex = 1;
  this.SearchKey = "";
  this.GetFirstPage = function () {
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    var self = this;
    wx.request({
      url: this.ApiUrl,
      data: {
        "PageInfo": {
          "PageIndex": this.PageIndex,
          "PageSize": 10
        },
        "Filter": {
          "SearchKey": this.SearchKey
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        self.RecvData(data);
      },
      error: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"获取视频");
    var data = recvData.data;
    if (this.Count <= 0) this.Count = data.count;                 //一共的视频个数
    if (data.list.length == 0) return;

    for (var i in data.list){
      var info = {};
      info.Addtime = data.list[i].addtime;
      info.Guid = data.list[i].guid;
      info.Imgurl = data.list[i].imgurl;
      info.Title = data.list[i].title;
      info.Videourl = data.list[i].videourl;

      this.Data.push(info);
    }
    
    if (this.Callback) this.Callback(this);
  }

  this.RecvError = function (request) {
    console.log("[StockReport::RecvError] ", request)
  }

  this.GetNextPage = function () {
    this.PageIndex++;
    var self = this;

    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.ApiUrl,
        data: {
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": 10
          },
          "Filter": {
            "SearchKey": this.SearchKey
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        error: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}

GetVedioList.SetDomain = function (domain, cacheDomain) {
  if (domain) g_VTradeResource.Domain = domain;
}

//导出统一使用JSCommon命名空间名
module.exports =
  {
    JSCommonVedio:
      {
        GetVedioList: GetVedioList
      }
  };
