// hqchart/src/jscommon/umychart.wechat/umychart.apidraw.wechat.js
/*
    远程API画图 返回图片
*/


function JSApiDrawResource() 
{
  this.Domain = "https://opensource.zealink.com";               //API域名
  this.CacheDomain = "https://opensourcecache.zealink.com";     //缓存域名
  this.CDNCacheDomain ="https://opensourcecdn.zealink.com";             //OSS回源CDN域名
}

var g_JSApiDrawResource = new JSApiDrawResource();

function JSApiDraw()
{
    
}

JSApiDraw.SetDomain = function (domain, cacheDomain, cdnDomain) 
{
  if (domain) g_JSAnalyzeResource.Domain = domain;
  if (cacheDomain) g_JSAnalyzeResource.CacheDomain = cacheDomain;
  if (cdnDomain) g_JSAnalyzeResource.CDNCacheDomain;
}

//分析工具类型枚举
var APIDRAW_TYPE=
{
    CHINAMAP_ID: 1  //中国地图
}

JSApiDraw.GetApiDrawControl=function(type)
{
    let apiDrawMap = new Map(
    [
        //类工厂
        [APIDRAW_TYPE.CHINAMAP_ID, { Create: function () { return new ApiDrawChinaMap() }, Name: '中国地图' }],
    ]);

    if (!apiDrawMap.has(type)) return null;

    return apiDrawMap.get(type).Create();
}


//画图基类
function IAPIDraw()
{
    this.Base64Data;    //图片数据
    this.ApiUrl;        //地址
    this.Callback;      //UI回调

    this.RequestData=function(data)
    {

    }

    this.RecvError=function(reqeust)
    {

    }
}


/*
    中国地图
    请求数据数据格式
    "data":
    [ 
        { "name":"安徽", "color":"#388bcf", "value":40 },
        { "name":"西藏", "color":"#d93b34", "value":50}
    ]
*/

function ApiDrawChinaMap()
{
    this.newMethod=IAPIDraw;   //派生
    this.newMethod();
    delete this.newMethod;

    this.ApiUrl=g_JSApiDrawResource.Domain+'/API/DrawChinaMap';

    this.RequestData=function(mapData)
    {
        var self=this;
        let reqeustData = { 'base64': true, data: mapData};
        wx.request({
            url: this.ApiUrl,
            data: reqeustData,
            method:'POST',
            dataType: 'json',
            success: function (data) 
            {
                self.RecvData(data);
            },
            fail: function (request) 
            {
                self.RecvError(request);
            }
        });
    }

    this.RecvData=function(recvData)
    {
        let data=recvData.data;
        if (data.base64data) this.Base64Data=data.base64data;
        if (this.Callback) this.Callback(this);
    }
}







//导出统一使用JSCommon命名空间名
module.exports=
{
    JSCommonApiDraw:
    {
        JSApiDraw:JSApiDraw,
        APIDRAW_TYPE:APIDRAW_TYPE,
    }
};
