// hqchart/src/jscommon/umychart.wechat/umychart.bigV.wechat.js
function JSNewsResource() {
  this.Domain = "https://apitest.zealink.com"; //API域名 opensource.zealink.com apitest.zealink.com
  this.CacheDomain = "https://opensourcecache.zealink.com"; //缓存域名
  this.JavaDomain = "https://javaweb7.zealink.com"; //java  API域名
}

var g_JSNewsResource = new JSNewsResource();

//大v观点
function GetBigVData() {
  this.UserId = '';
  this.Count; //总数
  this.CacheCount; //已下载的个数
  this.Type = 10;
  this.PageIndex = 1;
  this.PageSize = 20;
  this.SearchKey = "";
  this.StartDate = "";
  this.EndDate = "";

  this.Url = g_JSNewsResource.Domain + '/api/V_BigVList';          //NewsBigVListByES
  // this.Url = 'https://apitest.zealink.com/api/V_BigVList';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "Type": this.Type,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        },
        "Filter": {
          "SearchKey": this.SearchKey,
          "QueryDate": {
            "StartDate": this.StartDate,
            "EndDate": this.EndDate
          }
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"大v观点recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.DateTime = item.date;
      info.Level = item.level;
      info.Link = item.link;
      info.Newsid = item.newsid;
      info.Newstype = item.newstype;
      info.Picturelist = item.picturelist;
      info.Source = item.source;
      info.Stocklist = item.stocklist;
      info.Title = item.title;
      info.Isattention = item.isattention;
      info.Iscollection = item.iscollection;
      this.Data.push(info);
    }


    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "Type": this.Type,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          },
          "Filter": {
            "SearchKey": this.SearchKey,
            "QueryDate": {
              "StartDate": this.StartDate,
              "EndDate": this.EndDate
            }
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}


//大v观点-----加标签后的新接口
function GetBigVNewlist() {
  this.UserId = '';
  this.Count; //总数
  this.CacheCount; //已下载的个数
  this.Type = 10;
  this.PageIndex = 1;
  this.PageSize = 20;
  this.SearchKey = "";
  this.StartDate = "";
  this.EndDate = "";
  this.LabelInfo = {  //标签信息
    LabelGuid: "",
    LabelName: "",
    Type: 10,
    SearchType: 10,
    Sort: 3
  }; 


  this.Url = g_JSNewsResource.Domain + '/api/V_BigVList_V2';
  // this.Url = 'https://apitest.zealink.com/api/V_BigVList_V2';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "LabelInfo": {
          "LabelGuid": this.LabelInfo.LabelGuid,
          "LabelName": this.LabelInfo.LabelName,
          "Type": this.LabelInfo.Type,
          "SearchType": this.LabelInfo.SearchType,
          "Sort": this.LabelInfo.Sort
        },
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        },
        "Filter": {
          "SearchKey": this.SearchKey,
          "QueryDate": {
            "StartDate": this.StartDate,
            "EndDate": this.EndDate
          }
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"大v观点recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.DateTime = item.date;
      info.Level = item.level;
      info.Link = item.link;
      info.Newsid = item.newsid;
      info.Newstype = item.newstype;
      info.Picturelist = item.picturelist;
      info.Source = item.source;
      info.Stocklist = item.stocklist;
      info.Title = item.title;
      info.Isattention = item.isattention;
      info.Iscollection = item.iscollection;
      this.Data.push(info);
    }


    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "LabelInfo": {
            "LabelGuid": this.LabelInfo.LabelGuid,
            "LabelName": this.LabelInfo.LabelName,
            "Type": this.LabelInfo.Type,
            "SearchType": this.LabelInfo.SearchType,
            "Sort": this.LabelInfo.Sort
          },
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          },
          "Filter": {
            "SearchKey": this.SearchKey,
            "QueryDate": {
              "StartDate": this.StartDate,
              "EndDate": this.EndDate
            }
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}
GetBigVNewlist.init = function () {
  var news = new GetBigVNewlist();
  return news;
}


//大v观点详情页
function GetBigVDetail() {

  this.NewsId = '';

  this.Url = 'https://opensource.zealink.com/API/NewsBigVDetail';

  this.getBigVDetailData = function () {
    var self = this;
    //清空数据
    this.ViewDetail = {};

    wx.request({
      url: self.Url,
      data: {
        "newsid": self.NewsId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.ViewDetail.Content = recvData.data.content;
    this.ViewDetail.Date = recvData.data.date;
    this.ViewDetail.Source = recvData.data.source;
    this.ViewDetail.Title = recvData.data.title;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

}

GetBigVData.init = function () {
  var news = new GetBigVData();
  return news;
}
GetBigVDetail.init = function () {
  var newsDetail = new GetBigVDetail();
  return newsDetail;
}

//添加关注
function AddAttention(){
  this.UserId = '';
  this.Source = '';
  this.Newstype = 0;    //新闻类型
  this.Symbol = "";   
  this.Name = "";
  this.reqeustData = function(){
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_AddAttentionByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_AddAttentionByStock';
      var data = {
        "Symbol": this.Symbol,
        "Name": this.Name,
        "UserId": this.UserId
      };
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_AddAttention';   
      // this.Url = 'https://apitest.zealink.com/api/V_AddAttention';
      var data = {
        "WeChatSubscription": this.Source,
        "UserId": this.UserId
      };
    }
    
    var self = this;
  
    wx.request({
      url: this.Url,
      data: data,
      method: "POST",
      dataType: "json",
      success: function (data) {
        // console.log(data,"添加关注")
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  } 
}

AddAttention.init = function () {
  var news = new AddAttention();
  return news;
}

//取消关注
function CancelAttention() {
  this.UserId = '';
  this.Source = '';
  this.Newstype = 0;    //新闻类型
  this.Symbol = "";
  this.Name = "";

  this.reqeustData = function () {
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_CancelAttentionByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_CancelAttentionByStock';
      var data = {
        "Symbol": this.Symbol,
        "Name": this.Name,
        "UserId": this.UserId
      };
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_CancelAttention';   
      // this.Url = 'https://apitest.zealink.com/api/V_CancelAttention';
      var data = {
        "WeChatSubscription": this.Source,
        "UserId": this.UserId
      };
    }

    var self = this;

    wx.request({
      url: this.Url,
      data: data,
      method: "POST",
      dataType: "json",
      success: function (data) {
        // console.log(data, "添加关注")
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  }
}

CancelAttention.init = function () {
  var news = new CancelAttention();
  return news;
}

//收藏
function AddCollection() {
  this.UserId = '';
  this.Source = '';
  this.Newstype = 0;    //新闻类型  10是普通类新   20是调研类型
  this.NewsId = 0;      //新闻id

  this.Url = g_JSNewsResource.Domain + '/api/V_AddCollection';
  // this.Url = 'https://apitest.zealink.com/api/V_AddCollection';

  this.reqeustData = function () {
    var self = this;

    wx.request({
      url: this.Url,
      data: {
        "NewsType": this.Newstype,
        "NewsId": this.NewsId,
        "UserId": this.UserId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        // console.log(data, "添加收藏")
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  }
}
AddCollection.init = function () {
  var news = new AddCollection();
  return news;
}

//取消收藏
function CancelCollection() {
  this.UserId = '';
  this.Source = '';
  this.Newstype = 0;    //新闻类型  10是普通类新   20是调研类型
  this.NewsId = 0;      //新闻id

  this.Url = g_JSNewsResource.Domain + '/api/V_CancelCollection';
  // this.Url = 'https://apitest.zealink.com/api/V_CancelCollection';

  this.reqeustData = function () {
    var self = this;

    wx.request({
      url: this.Url,
      data: {
        "NewsType": this.Newstype,
        "NewsId": this.NewsId,
        "UserId": this.UserId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        // console.log(data, "添加收藏")
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  }
}
CancelCollection.init = function () {
  var news = new CancelCollection();
  return news;
}

//拉黑
function AddBlackList() {
  this.UserId = '';
  this.Source = '';
  this.Newstype = 0;    //新闻类型
  this.Symbol = "";
  this.Name = "";

  this.reqeustData = function () {
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_AddBlacklistByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_AddBlacklistByStock';
      var data = {
        "Symbol": this.Symbol,
        "Name": this.Name,
        "UserId": this.UserId
      };
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_AddBlacklist';
      // this.Url = 'https://apitest.zealink.com/api/V_AddBlacklist';
      var data = {
        "WeChatSubscription": this.Source,
        "UserId": this.UserId
      };
    }

    var self = this;

    wx.request({
      url: this.Url,
      data: data,
      method: "POST",
      dataType: "json",
      success: function (data) {
        // console.log(data,"拉黑")
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  }
}

AddBlackList.init = function () {
  var news = new AddBlackList();
  return news;
}

//取消拉黑
function CancelBlackList() {
  this.UserId = '';
  this.Source = '';
  this.Newstype = 0;    //新闻类型
  this.Symbol = "";
  this.Name = "";

  this.reqeustData = function () {
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_CancelBlacklistByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_CancelBlacklistByStock';
      var data = {
        "Symbol": this.Symbol,
        "Name": this.Name,
        "UserId": this.UserId
      };
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_CancelBlacklist';
      // this.Url = 'https://apitest.zealink.com/api/V_CancelBlacklist';
      var data = {
        "WeChatSubscription": this.Source,
        "UserId": this.UserId
      };
    }

    var self = this;

    wx.request({
      url: this.Url,
      data: data,
      method: "POST",
      dataType: "json",
      success: function (data) {
        // console.log(data,"拉黑")
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  }
}

CancelBlackList.init = function () {
  var news = new CancelBlackList();
  return news;
}


//关注列表
function GetAttentionData() {
  this.UserId = '';
  this.Count; //总数
  this.CacheCount; //已下载的个数
  this.Type = 10;
  this.PageIndex = 1;
  this.PageSize = 20;
  this.SearchKey = "";
  this.StartDate = "";
  this.EndDate = "";

  this.Url = g_JSNewsResource.Domain + '/api/V_AttentionList';
  // this.Url = 'https://apitest.zealink.com/api/V_AttentionList';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "Type": this.Type,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        },
        "Filter": {
          "SearchKey": this.SearchKey,
          "QueryDate": {
            "StartDate": this.StartDate,
            "EndDate": this.EndDate
          }
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"大v观点recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.DateTime = item.date;
      info.Level = item.level;
      info.Link = item.link;
      info.Newsid = item.newsid;
      info.Newstype = item.newstype;
      info.Picturelist = item.picturelist;
      info.Source = item.source;
      info.Stocklist = item.stocklist;
      info.Title = item.title;
      info.Isattention = item.isattention;
      info.Iscollection = item.iscollection;
      this.Data.push(info);
    }


    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "Type": this.Type,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          },
          "Filter": {
            "SearchKey": this.SearchKey,
            "QueryDate": {
              "StartDate": this.StartDate,
              "EndDate": this.EndDate
            }
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}

GetAttentionData.init = function () {
  var news = new GetAttentionData();
  return news;
}

//我的关注
function GetMyAttentionList(){
  this.UserId = '';
  this.PageIndex = 1;
  this.PageSize = 20;
  this.Newstype = 0;    //新闻类型  20是调研 10是普通新闻类型

  this.Count; //总数
  this.CacheCount; //已下载的个数
  
  this.getFirstData = function(){
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyAttentionListByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_MyAttentionListByStock';
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyAttentionList';
      // this.Url = 'https://apitest.zealink.com/api/V_MyAttentionList';
    }

    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"我的关注recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    // if (this.Count <= 0) this.Count = recvData.data.count; //一共个数
    this.Count = recvData.data.count; //一共个数

    if (this.Newstype == 20){
      for (var i in recvData.data.list){
        var item = recvData.data.list[i];
        var info = {};
        info.Name = item.name;
        info.Symbol = item.symbol;
        this.Data.push(info);
      }
    }else{
      for (var i in recvData.data.list) {
        var item = recvData.data.list[i];
        this.Data.push(item);
      }
    }

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyAttentionListByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_MyAttentionListByStock';
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyAttentionList';
      // this.Url = 'https://apitest.zealink.com/api/V_MyAttentionList';
    }

    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}
GetMyAttentionList.init = function () {
  var news = new GetMyAttentionList();
  return news;
}

//我的收藏
function GetMyCollectionList(){
  this.UserId = '';
  this.PageIndex = 1;
  this.PageSize = 20;
  this.Count; //总数
  this.CacheCount; //已下载的个数

  this.Url = g_JSNewsResource.Domain + '/api/V_CollectionList';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"我的收藏recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.DateTime = item.date;
      info.Level = item.level;
      info.Link = item.link;
      info.Newsid = item.newsid;
      info.Newstype = item.newstype;
      info.Picturelist = item.picturelist;
      info.Source = item.source;
      info.Stocklist = item.stocklist;
      info.Title = item.title;
      info.Isattention = item.isattention;
      info.Iscollection = item.iscollection;
      this.Data.push(info);
    }

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }

}
GetMyCollectionList.init = function () {
  var news = new GetMyCollectionList();
  return news;
}

//我的黑名单
function GetMyBlackList() {
  this.UserId = '';
  this.PageIndex = 1;
  this.PageSize = 20;
  this.Newstype = 0;    //新闻类型  20是调研 10是普通新闻类型

  this.Count; //总数
  this.CacheCount; //已下载的个数

  this.getFirstData = function () {
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyBlacklistListByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_MyBlacklistListByStock';
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyBlacklistList';
      // this.Url = 'https://apitest.zealink.com/api/V_MyBlacklistList';
    }

    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"我的关注recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    // if (this.Count <= 0) this.Count = recvData.data.count; //一共个数
    this.Count = recvData.data.count;

    if (this.Newstype == 20) {
      for (var i in recvData.data.list) {
        var item = recvData.data.list[i];
        var info = {};
        info.Name = item.name;
        info.Symbol = item.symbol;
        this.Data.push(info);
      }
    } else {
      for (var i in recvData.data.list) {
        var item = recvData.data.list[i];
        this.Data.push(item);
      }
    }

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    if (this.Newstype == 20) {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyBlacklistListByStock';
      // this.Url = 'https://apitest.zealink.com/api/V_MyBlacklistListByStock';
    } else {
      this.Url = g_JSNewsResource.Domain + '/api/V_MyBlacklistList';
      // this.Url = 'https://apitest.zealink.com/api/V_MyBlacklistList';
    }

    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}
GetMyBlackList.init = function () {
  var news = new GetMyBlackList();
  return news;
}

//标签---获取标签列表:未授权
function GetDefalutLabel() {
  this.Url = g_JSNewsResource.Domain + '/api/V_GetDefalutLabel';
  // this.Url = 'https://apitest.zealink.com/api/V_GetDefalutLabel';

  this.reqeustData = function () {
    var self = this;
    //清空数据
    this.Data = [];

    wx.request({
      url: this.Url,
      data: {},
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }
  this.RecvData = function (recvData) {
    console.log(recvData.data,"未授权标签列表recvData")
    this.Data = recvData.data.list;

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
GetDefalutLabel.init = function () {
  var news = new GetDefalutLabel();
  return news;
}

//标签---获取标签列表:已授权
function GetAllIndex(){
  this.UserId = "";
  this.Url = g_JSNewsResource.Domain + '/api/V_LabelList';
  // this.Url = 'https://apitest.zealink.com/api/V_LabelList';

  this.reqeustData = function(){
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }
  this.RecvData = function (recvData) {
    // console.log(recvData.data,"标签列表recvData")
    this.Data = {
      Mylist:[],
      Otherlist:[]
    };

    if (recvData.data.mylist) this.Data.Mylist = recvData.data.mylist;
    if (recvData.data.otherlist) this.Data.Otherlist = recvData.data.otherlist;

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
GetAllIndex.init = function () {
  var news = new GetAllIndex();
  return news;
}

//标签---添加标签
function AddSelfLabel() {
  this.UserId = "";
  this.LabelGuid = "";
  this.Url = g_JSNewsResource.Domain + '/api/V_AddLabel';
  // this.Url = 'https://apitest.zealink.com/api/V_AddLabel';

  this.reqeustData = function () {
    var self = this;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "LabelGuid": this.LabelGuid
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        /*wx.showToast({
          title: data.data.message
        })
        return;*/
        self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  };
  this.RecvData = function (recvData) {
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
AddSelfLabel.init = function () {
  var news = new AddSelfLabel();
  return news;
}

//标签---删除标签
function DeleteSelfLabel() {
  this.UserId = "";
  this.LabelGuid = "";
  this.Url = g_JSNewsResource.Domain + '/api/V_DeleteLabel';
  // this.Url = 'https://apitest.zealink.com/api/V_DeleteLabel';

  this.reqeustData = function () {
    var self = this;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "LabelGuid": this.LabelGuid
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        /*wx.showToast({
          title: data.data.message
        })
        return;*/
        self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  };
  this.RecvData = function (recvData) {
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
DeleteSelfLabel.init = function () {
  var news = new DeleteSelfLabel();
  return news;
}

//标签---调整顺序
function ChangeLabelSort(){
  this.UserId = "";
  this.Url = g_JSNewsResource.Domain + '/api/V_LabelChangePosition';
  // this.Url = 'https://apitest.zealink.com/api/V_LabelChangePosition';
  this.Self = {
    LabelGuid: "",
    LabelName: "",
    Type: 10,
    SearchType: 10,
    Sort: 3
  };
  this.TargetPrev = {
    LabelGuid: "",
    LabelName: "",
    Type: 10,
    SearchType: 10,
    Sort: 3
  }

  this.reqeustData = function () {
    var self = this;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "Self": {
          "LabelGuid": this.Self.LabelGuid,
          "LabelName": this.Self.LabelName,
          "Type": this.Self.Type,
          "SearchType": this.Self.SearchType,
          "Sort": this.Self.Sort
        },
        "TargetPrev": {
          "LabelGuid": this.TargetPrev.LabelGuid,
          "LabelName": this.TargetPrev.LabelName,
          "Type": this.TargetPrev.Type,
          "SearchType": this.TargetPrev.SearchType,
          "Sort": this.TargetPrev.Sort
        },
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}

ChangeLabelSort.init = function () {
  var news = new ChangeLabelSort();
  return news;
}

//评论列表
function GetCommentList(){
  this.UserId = "";
  this.NewsId = "";   //文章id
  this.ParendId = "";   //评论父id
  this.Openid = "";      //微信openid
  this.PageIndex = 1;    //请求页码
  this.PageSize = 10;   //每页请求条数
  this.Url = g_JSNewsResource.JavaDomain + '/comments/list';

  this.getFirstData = function(){
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "userId": this.UserId,
        "sid": this.NewsId,
        "pid": this.ParendId,
        "openid": this.Openid,
        "page": 1,
        "rows": this.PageSize
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data.data,"评论")
    if (recvData.data.data.records.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.data.counts; //一共个数

    for (var i in recvData.data.data.records){
      var item = recvData.data.data.records[i];
      var info = {};
      info.Content = item.content;
      info.CreateTime = item.create_time;
      info.Avatar = item.avatar;     //头像
      info.Nickname = item.nickname;
      info.ID = item.id;
      info.LoveStatus = item.love_status;     // 表示当前用户是否赞过该评论;
      info.UserId = item.user_id;       // 用户id;
      info.LoveCount = item.love_count; // 赞
      info.DisloveCount = item.dislove_count; // 踩

      this.Data.push(info);
    }
    
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "userId": this.UserId,
          "sid": this.NewsId,
          "pid": this.ParendId,
          "openid": this.Openid,
          "page": this.PageIndex,
          "rows": this.PageSize
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }


}
GetCommentList.init = function () {
  var news = new GetCommentList();
  return news;
}

//发表评论
function  ReportComment(){
  this.UserId = "";
  this.NewsId = "";   //文章id
  this.ParendId = "";   //评论父id
  this.Openid = "";      //微信openid
  this.Content = "";     //评论内容
  this.Nickname = "" ;   //评论人昵称
  this.Avatar = "";      //评论人头像
  this.Url = g_JSNewsResource.JavaDomain + '/comments/save';

  this.requestData = function(){
    var self = this;
    //清空数据
    this.Data = [];

    wx.request({
      url: this.Url,
      data: {
        "userId": this.UserId,
        "sid": this.NewsId,
        "pid": this.ParendId,
        "openid": this.Openid,
        "content": this.Content,
        "nickname": this.Nickname,
        "avatar": this.Avatar
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
ReportComment.init = function () {
  var news = new ReportComment();
  return news;
}

//删除评论
function DeleteComment() {
  this.CommentId = "";     //该条评论的id
  // this.Url = g_JSNewsResource.JavaDomain + '/comments/delete';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    wx.request({
      url: g_JSNewsResource.JavaDomain + '/comments/delete/' + this.CommentId,
      // data: {
      //   "id": this.CommentId
      // },
      method: "GET",
      dataType: "json",
      success: function (data) {
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
DeleteComment.init = function () {
  var news = new DeleteComment();
  return news;
}

//点赞和取消点赞
function IsSupport() {
  this.UserId = "";
  this.CommentId = "";     //该条评论的id
  this.Url = g_JSNewsResource.JavaDomain + '/comments/love';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = [];

    wx.request({
      url: this.Url,
      data: {
        "userId": this.UserId,
        "bizId": this.CommentId,
        "bizType": 2          //1.文章，2.评论'，默认填2
      },
      method: "GET",
      dataType: "json",
      success: function (data) {
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.Data = recvData.data;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
IsSupport.init = function () {
  var news = new IsSupport();
  return news;
}

//踩和取消踩
function IsOppose() {
  this.UserId = "";
  this.CommentId = "";     //该条评论的id
  this.Url = g_JSNewsResource.JavaDomain + '/comments/dislove';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = [];

    wx.request({
      url: this.Url,
      data: {
        "userId": this.UserId,
        "bizId": this.CommentId,
        "bizType": 2          //1.文章，2.评论'，默认填2
      },
      method: "GET",
      dataType: "json",
      success: function (data) {
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
IsOppose.init = function () {
  var news = new IsOppose();
  return news;
}

// 判断是否需要隐藏标签;
function IsHiddenTabBar() {
  this.Value = 1;
  this.Product = 50;
  this.Url = g_JSNewsResource.Domain + '/api/TabBar_ShowFlash';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = [];

    wx.request({
      url: this.Url,
      data: {
        "Value": this.Value,
        "Product": this.Product,
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.Data = recvData.data;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
IsHiddenTabBar.init = function () {
  var news = new IsHiddenTabBar();
  return news;
}

//单个公众号新闻列表、单个股票的调研列表
function GetSingleList(){
  this.Count; //总数
  this.CacheCount; //已下载的个数
  this.UserId = "";
  this.NewsType = 10;
  this.SingleKey = "";
  this.PageIndex = 1;
  this.PageSize = 20;
  this.SearchKey = "";
  this.StartDate = ""; 
  this.EndDate = "";
  this.Url = g_JSNewsResource.Domain + '/api/V_BigVListBySingle';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "NewsType": this.NewsType,
        "SingleKey": this.SingleKey,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        },
        "Filter": {
          "SearchKey": this.SearchKey,
          "QueryDate": {
            "StartDate": this.StartDate,
            "EndDate": this.EndDate
          }
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"单个公众号recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.DateTime = item.date;
      info.Level = item.level;
      info.Link = item.link;
      info.Newsid = item.newsid;
      info.Newstype = item.newstype;
      info.Picturelist = item.picturelist;
      info.Source = item.source;
      info.Stocklist = item.stocklist;
      info.Title = item.title;
      info.Isattention = item.isattention;
      info.Iscollection = item.iscollection;
      this.Data.push(info);
    }


    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.UserId,
          "NewsType": this.NewsType,
          "SingleKey": this.SingleKey,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          },
          "Filter": {
            "SearchKey": this.SearchKey,
            "QueryDate": {
              "StartDate": this.StartDate,
              "EndDate": this.EndDate
            }
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}
GetSingleList.init = function () {
  var news = new GetSingleList();
  return news;
}

// 公众号联想功能
function Associate(){
  this.SearchKey = "";
  this.PageIndex = 1;
  this.PageSize = '';
  this.Url = g_JSNewsResource.Domain + '/api/V_SourceSearch';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "SearchKey": this.SearchKey,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        },
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"单个公众号recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    this.Data = recvData.data.list;

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "SearchKey": this.SearchKey,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          },
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}
Associate.init = function () {
  var news = new Associate();
  return news;
}

// 股票联想功能
function StockAssociate(){
  this.Input = "";
  this.Start = 0;
  this.End = '';
  this.Type = 0;
  this.Url = 'https://opensource.zealink.com/API/StockSpell';

  this.getFirstData = function () {
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "input": this.Input,
        "start": this.Start,
        "end": this.End,
        "type": this.Type
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData,"单个公众号recvData")
    if (recvData.data.stock.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    this.Data = recvData.data.stock;

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function () {
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

    if (this.PageIndex <= totalPages) {
      wx.request({
        url: this.Url,
        data: {
          "input": this.Input,
          "start": this.Start,
          "end": this.End,
          "type": this.Type
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    } else {
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}
StockAssociate.init = function () {
  var news = new StockAssociate();
  return news;
};

// 是否有管理权限;
function Ismanagement() {
  this.UserId = '';
  this.Url = g_JSNewsResource.Domain + '/api/V_IsDeletePermission';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = {};

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        console.log(data, 'data');
        // if (data.data.code != 0) {
        //   wx.showModal({
        //     title: '提示',
        //     content: data.data.message,
        //   })
        //   return;
        // }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.Data = recvData.data;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
Ismanagement.init = function () {
  var news = new Ismanagement();
  return news;
}

// 删除广告;
function DeleteAds() {
  this.UserId = '';
  this.NewsId = '';
  this.Url = 'https://projectapi-web4.zealink.com/conf/api/DeleteBigV';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = {};

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "NewsId": this.NewsId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        console.log(data, 'data');
        /*if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }*/
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.Data = recvData.data;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
DeleteAds.init = function () {
  var news = new DeleteAds();
  return news;
}

// 取消推荐;
function CancelRecommand() {
  this.UserId = '';
  this.NewsId = '';
  this.Url = 'https://projectapi-web4.zealink.com/conf/api/CancelRecommendBigV';

  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = {};

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "NewsId": this.NewsId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        console.log(data, 'data');
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.Data = recvData.data;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
CancelRecommand.init = function () {
  var news = new CancelRecommand();
  return news;
};

// 添加推荐;
function AddRecommand() {
  this.UserId = '';
  this.NewsId = '';
  this.Url = 'https://projectapi-web4.zealink.com/conf/api/RecommendBigV';
  this.requestData = function () {
    var self = this;
    //清空数据
    this.Data = {};

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.UserId,
        "NewsId": this.NewsId
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        console.log(data, 'data');
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    this.Data = recvData.data;
    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }
}
AddRecommand.init = function () {
  var news = new AddRecommand();
  return news;
}

module.exports = {
  JSCommonVPoint: {
    GetBigVNewlist: GetBigVNewlist,   //大v观点-----加标签后的新接口
    GetBigVData: GetBigVData,
    GetBigVDetail: GetBigVDetail,        
    AddAttention: AddAttention,          //添加关注
    CancelAttention: CancelAttention,         //取消关注
    AddCollection: AddCollection,            //添加收藏
    CancelCollection: CancelCollection,       //取消收藏
    AddBlackList: AddBlackList,            //拉黑
    CancelBlackList: CancelBlackList,      //取消拉黑
    GetAttentionData: GetAttentionData,       //关注列表
    GetMyAttentionList: GetMyAttentionList,     //我的关注--系统
    GetMyCollectionList: GetMyCollectionList,     //我的收藏--系统
    GetMyBlackList: GetMyBlackList,         //我的黑名单--系统
    GetDefalutLabel: GetDefalutLabel,    //标签--- 获取标签列表:未授权
    GetAllIndex: GetAllIndex,          //标签--- 获取标签列表:已授权
    AddSelfLabel: AddSelfLabel,          //标签---添加标签
    DeleteSelfLabel: DeleteSelfLabel,     //标签---删除标签
    ChangeLabelSort: ChangeLabelSort,       //标签---改变顺序
    GetCommentList: GetCommentList,        //评论列表
    ReportComment: ReportComment,          //发表评论
    DeleteComment: DeleteComment,     //删除评论
    IsSupport: IsSupport,       // 点赞和取消点赞  
    IsOppose: IsOppose,         //踩和取消踩 
    GetSingleList: GetSingleList,  //单个公众号新闻列表、单个股票的调研列表,
    IsHiddenTabBar: IsHiddenTabBar,
    Associate: Associate,    // 公众号联想功能
    StockAssociate: StockAssociate,
    Ismanagement: Ismanagement,    // 是否有管理权限
    DeleteAds: DeleteAds,           // 删除广告
    CancelRecommand: CancelRecommand,    // 取消推荐
    AddRecommand: AddRecommand            // 添加推荐
  }
};
