// hqchart/src/jscommon/umychart.wechat/umychart.news.wechat.js
/*
 微信接口的个股新闻
 */
import {
    JSCommonDate
} from "getDate.wechat.js"; //获取时间

function JSNewsResource() {
    this.Domain = "https://opensource.zealink.com"; //API域名 opensource.zealink.com apitest.zealink.com
    this.CacheDomain = "https://opensourcecache.zealink.com"; //缓存域名
    this.JavaDomain = "https://javaweb7.zealink.com"; //java  API域名
}

var g_JSNewsResource = new JSNewsResource();


function JSNews() {

}

JSNews.SetDomain = function (domain, cacheDomain) {
    if (domain) g_JSNewsResource.Domain = domain;
    if (cacheDomain) g_JSNewsResource.CacheDomain = cacheDomain;
}

//获取新闻列表
JSNews.GetNews = function (symbol, newsType) {
    switch (newsType) {
        case NEWS_TYPE.STOCK_NEWS:
            return new StockNews(symbol);
        case NEWS_TYPE.ANNOUNCEMENT_ANALYSE:
            return new AnnouncementAnalyse(symbol);
        case NEWS_TYPE.STOCK_INTERACT:
            return new StockInteract(symbol);
        case NEWS_TYPE.MARKET_SEARCH:
            return new MarketSearch(symbol);
        case NEWS_TYPE.STOCK_NEGATIVE:
            return new StockNegative(symbol);
        case NEWS_TYPE.STOCK_REPORT:
            return new StockReport(symbol);
        case NEWS_TYPE.NEW_STOCK_NEGATIVE:
            return new newStockReport(symbol);
        case NEWS_TYPE.OFFICIAL_WEBSITE_NEWS:
            return new OfficialWebsiteNews(symbol);
        case NEWS_TYPE.SENIOR_EXECUTIVE_NEWS:
            return new SeniorExecutiveNews(symbol);
        case NEWS_TYPE.VIEWPOINTS_BY_VIP:
            return new ViewpointsByVIP(symbol);
        case NEWS_TYPE.TOP_MANAGERS_RISK_NEWS:
            return new TopManagersRiskNews(symbol);
        case NEWS_TYPE.FAILURE_REGROUP_NEWS:
            return new FailureRegroupNews(symbol);
        case NEWS_TYPE.TOP_MANAGERS_RISK_LIST:
            return new TopManagersRiskList(symbol);
        case NEWS_TYPE.SEARCH_NOTICE:
            return new SearchNotice(symbol);
    }
}
//获取新闻内容
JSNews.GetContent = function (id, newsType) {
    switch (newsType) {
        case NEWS_TYPE.STOCK_NEWS:
            return new StockNewsContent(id);
            break;
        case NEWS_TYPE.MARKET_SEARCH:
            return new marketSearchInfo(id);
            break;
        case NEWS_TYPE.NEW_STOCK_NEGATIVE:
            return new stockReportInfo(id);
            break;

        case NEWS_TYPE.OFFICIAL_WEBSITE_NEWS:
            return new OfficialWebsiteInfo(id);
            break;
        case NEWS_TYPE.SENIOR_EXECUTIVE_NEWS:
            return new SeniorExecutiveInfo(id);
            break;
        case NEWS_TYPE.VIEWPOINTS_BY_VIP:
            return new ViewpointsByVIPInfo(id);
            break;
        case NEWS_TYPE.TOP_MANAGERS_RISK_NEWS:
            return new TopManagersRiskNewsInfo(id);
            break;
    }
}

//新闻类型
var NEWS_TYPE = {
    STOCK_NEWS: 1, //个股新闻
    ANNOUNCEMENT_ANALYSE: 2, //公告解读
    STOCK_INTERACT: 3, //互动易
    MARKET_SEARCH: 4, //调研
    STOCK_NEGATIVE: 5, //负面新闻
    STOCK_REPORT: 6, //公告
    NEW_STOCK_NEGATIVE: 7, //新的负面新闻接口 NewsAllByES
    OFFICIAL_WEBSITE_NEWS: 8, //官网新闻
    SENIOR_EXECUTIVE_NEWS: 9, //高管新闻
    VIEWPOINTS_BY_VIP: 10, //大V观点新闻
    TOP_MANAGERS_RISK_NEWS: 11, //高管风险新闻
    FAILURE_REGROUP_NEWS: 12, //重组失败
    TOP_MANAGERS_RISK_LIST: 13, //高管风险类型新闻
    SEARCH_NOTICE: 14, //公告搜索（uts/服务/公告）
}

//新闻资讯基类
function INewsBase(symbol) {
    this.Symbol = symbol; //股票代码
    this.Name; //股票名称
    this.Callback; //回调 function(info,news) info: {Start:请求的新闻起始位置 , End:新闻的结束位置 } , news 对应的 新闻类

    this.Data = new Array(); //新闻数据  { Date:日期, Title:标题 , Source:来源, ID:新闻ID}
    this.Count; //一共的新闻个数
    this.CacheCount; //已下载的新闻个数
    this.PageSize = 10; //每页的新闻个数
    this.ApiUrl; //数据请求地址

    this.StartDate; //开始时间
    this.SndDate; //结束时间

    this.Aggregate = 10; //统计排名

    //设置默认的查询区间
    this.SetDefautQueryDate = function () {
        // 取最近1年的数据
        let date = new Date();
        this.StartDate = (date.getFullYear() - 1) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
        this.EndDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
    }

    //把查询的股票统一转化为数组
    this.GetQuerySymbol = function () {
        let arySymbol = new Array();
        if (typeof(this.Symbol) == 'string') arySymbol.push(this.Symbol); //单个股票
        else arySymbol = this.Symbol.slice(0);

        return arySymbol;
    }

    //是否是最后一页
    this.IsEndPage = function () {
        if (this.Count == null || this.Count <= 0) return true;

        return this.CacheCount >= this.Count;

    }

    //获取第1页
    this.GetFirstPage = function () {

    }

    //下拉获取下一页
    this.GetNextPage = function () {

    }
}

//新闻详情基类
function INewsContent(id) {
    this.ID = id; //新闻id
    this.Symbol; //股票代码
    this.Name; //股票名称
    this.ApiUrl;
    this.Callback; //回调 function(this)
    this.Data; //数据 { Symbol:代码, Name:名称, Title:标题, Content:内容, Date:日期 .....}
    this.Error; //如果调用失败 把错误信息写这里

    this.GetContent = function () {

    }
}

var RECV_DATA_TYPE = {
    NEWS_DATA: 1, //个股新闻数据
}


//个股新闻
function StockNews(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    var date = new Date();
    this.EndDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate(); //获取结束时间
    this.StartDate = (date.getFullYear() - 1) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsStockListByES';

    var self = this;

    //获取第1页
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;
        let arySymbol = this.GetQuerySymbol();
        wx.request({
            url: this.ApiUrl,
            data: {
                "filter": {
                    "querydate": {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    "symbollist": arySymbol,
                    "searchkey": null,
                },
                "pageinfo": {
                    "pageindex": 1,
                    "pagesize": this.PageSize
                },
                "userid": null
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                this.RecvError(request);
            }
        })
    }


    this.RecvData = function (recvData) {
        console.log("个股新闻", recvData)
        var data = recvData.data;
        if (this.Count <= 0)
            this.Count = data.count;

        var info = {};
        info.Start = this.Data.length;
        info.End = this.Data.length;

        for (var i in data.list) {
            var item = data.list[i];
            this.Data.push({
                ID: item.id,
                Name: item.name,
                Releasedate: item.date,
                Source: item.source,
                Symbol: item.symbol,
                Title: item.title
            })
            if (i > 0) ++info.End;
        }

        this.CacheCount = this.Data.length;
        this.InvokeCallback(info); //通知页面
    }

    this.RecvError = function (request) {
        console.log("[StockInteract::RecvError] ", request)
    }

    this.InvokeCallback = function (info) {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(info, this);
    }


    //下拉获取下一页
    this.GetNextPage = function () {
        var self = this;

        let pageIndex = (this.Data.length / this.PageSize) + 1;
        let arySymbol = this.GetQuerySymbol();

        wx.request({
            url: this.ApiUrl,
            data: {
                "filter": {
                    "querydate": {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    "symbollist": arySymbol,
                    "searchkey": null,
                },
                "pageinfo": {
                    "pageindex": pageIndex,
                    "pagesize": this.PageSize
                },
                "userid": null
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }
}

//调研
function MarketSearch(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.Data = {
        Search: [],
        Aggregate: []
    };

    var date = new Date();
    this.EndDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate(); //获取结束时间
    this.StartDate = (date.getFullYear() - 1) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

    this.ApiUrl = g_JSNewsResource.Domain + '/API/InvestorRelationsList';
    this.SearchKey = null;

    //获取第1页
    this.GetFirstPage = function () {
        var self = this;
        //清空数据
        this.Data = {
            Search: [],
            Aggregate: []
        };
        this.Count = 0;
        this.CacheCount = 0;

        if (typeof(this.Symbol) == "string") {
            var data = {
                "symbol": [this.Symbol],
                "filed": ["name", "symbol", "releasedate", "title", "id", "researchdate", "level", "type", "researcher"],
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "calccount": 1,
                "start": 0,
                "end": this.PageSize,
                "userid": "qiuyh"
            }
        } else {
            var data = {
                "symbol": this.Symbol,
                "filed": ["name", "symbol", "releasedate", "title", "id", "researchdate", "level", "type", "researcher"],
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": this.SearchKey,
                "calccount": 1,
                "start": 0,
                "end": this.PageSize,
                "userid": "qiuyh",
                "aggregate": this.Aggregate
            }
        }
        wx.request({
            url: this.ApiUrl,
            data: data,
            method: "POST",
            dataType: "json",
            success: function (data) {
                if (data.data.code != 0) {
                    wx.showModal({
                        title: '提示',
                        content: data.data.message,
                    })
                    return;
                }
                self.RecvData(data);
            },
            fail: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log("调研", recvData)
        var data = recvData.data;
        if (this.Count <= 0)
            this.Count = data.count;

        var info = {};
        info.Start = this.Data.length;
        info.End = this.Data.length;

        if (data.aggregate) {
            for (var i in data.aggregate) {
                this.Data.Aggregate.push({
                    Name: data.aggregate[i].value2,
                    Symbol: data.aggregate[i].value,
                    Count: data.aggregate[i].count
                });
            }
            // this.Data.Aggregate = data.aggregate;
        }

        for (var i in data.list) {
            var item = data.list[i];
            this.Data.Search.push({
                ID: item.id,
                Name: item.name,
                Symbol: item.symbol,
                Releasedate: item.releasedate,
                Researchdate: item.researchdate,
                Title: item.title,
                Type: item.type,
                Researcher: item.researcher,
                Level: item.level
            })
            if (i > 0) ++info.End;
        }

        this.CacheCount = this.Data.Search.length;
        this.InvokeCallback(info); //通知页面
    }

    this.RecvError = function (request) {
        console.log("[StockInteract::RecvError] ", request)
    }

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        var info = {};
        this.Callback(info, this);
    }

    //下拉获取下一页
    this.GetNextPage = function () {
        var self = this;

        var start = this.Data.Search.length;
        var end = start + this.PageSize;

        var data = {
            "symbol": [this.Symbol],
            "filed": ["name", "symbol", "releasedate", "title", "id", "researchdate", "level", "type", "researcher"],
            "querydate": {
                "StartDate": this.StartDate,
                "EndDate": this.EndDate
            },
            "calccount": 1,
            "start": start,
            "end": end,
            "userid": "qiuyh"
        };

        var start = this.Data.Search.length;
        var end = start + this.PageSize;

        if (typeof(this.Symbol) == "string") {
            var data = {
                "symbol": [this.Symbol],
                "filed": ["name", "symbol", "releasedate", "title", "id", "researchdate", "level", "type", "researcher"],
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "calccount": 1,
                "start": start,
                "end": end,
                "userid": "qiuyh"
            };
        } else {
            var data = {
                "symbol": this.Symbol,
                "filed": ["name", "symbol", "releasedate", "title", "id", "researchdate", "level", "type", "researcher"],
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": this.SearchKey,
                "calccount": 1,
                "start": start,
                "end": end,
                "userid": "qiuyh",
                "aggregate": this.Aggregate
            }
        }
        if (this.Data.Search.length < this.Count) {
            wx.request({
                url: this.ApiUrl,
                data: data,
                method: "POST",
                dataType: "json",
                success: function (data) {
                    if (data.data.code != 0) {
                        wx.showModal({
                            title: '提示',
                            content: data.data.message,
                        })
                        return;
                    }
                    self.RecvData(data);
                },
                error: function (request) {
                    this.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }

    }
}


//个股公告分析
function AnnouncementAnalyse(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/AnnualReportAllType';
    this.Error = null;

    this.GetFirstPage = function () {
        this.Error = null; //清空错误信息
        this.Data = [];
        this.Name = null;

        var self = this;
        var date = new Date();

        // 取最近3年的数据
        var start = (date.getFullYear() - 3) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
        var end = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": this.Symbol,
                "querydate": {
                    "StartDate": start,
                    "EndDate": end
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request)
            }
        })
    }

    this.RecvData = function (recvData) {
        //console.log(recvData);
        var data = recvData.data;
        if (data.list == null) return this.InvokeCallback();

        for (var i in data.list) {
            var item = data.list[i];
            if (item == null || item.info == null || item.info.length <= 0) continue;

            var analyseData = {
                Name: item.typename,
                Data: new Array()
            };

            for (var j in item.info) {
                var subItem = item.info[j];
                analyseData.Data.push({
                    Date: subItem.date,
                    Content: subItem.abstract
                });

                if (this.Name == null) this.Name = subItem.name;
            }

            this.Data.push(analyseData);
        }

        this.InvokeCallback(); //通知页面
    }

    this.RecvError = function (request) {
        console.log('[AnnouncementAnalyse:RecvError] ', "request");
        this.Error = "请求失败";
    }

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        var info = {};
        this.Callback(info, this);
    }

}

//互动易
function StockInteract(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsInteract';

    this.Data = {
        News: [],
        Aggregate: []
    };
    this.SearchKey = "";
    this.userid = null;


    //获取第1页
    this.GetFirstPage = function () {
        //清空数据
        this.Data = {
            News: [],
            Aggregate: []
        };
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        // 取最近1年的数据
        // var date = new Date();
        // var startDate = (date.getFullYear() - 1) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
        // var endDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

        if (typeof(this.Symbol) == "string") {
            var data = {
                "symbol": [this.Symbol],
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": this.SearchKey,
                "calccount": 1, //返回一共的数据个数
                "start": 0,
                "end": this.PageSize,
                "userid": this.userid
            }
        } else {
            var data = {
                "symbol": this.Symbol,
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": this.SearchKey,
                "calccount": 1, //返回一共的数据个数
                "start": 0,
                "end": this.PageSize,
                "userid": this.userid,
                "aggregate": this.Aggregate
            }
        }

        wx.request({
            url: this.ApiUrl,
            data: data,
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log("recvDatainteract", recvData.data.list)
        var data = recvData.data;
        if (this.Count <= 0)
            this.Count = data.count; //一共的新闻个数

        var info = {};
        info.Start = this.Data.length;
        info.End = this.Data.length;

        if (data.aggregate) {
            for (var i in data.aggregate) {
                this.Data.Aggregate.push({
                    Name: data.aggregate[i].value2,
                    Symbol: data.aggregate[i].value,
                    Count: data.aggregate[i].count
                });
            }
            // this.Data.Aggregate = data.aggregate;
        }

        for (var i in data.list) {
            var item = data.list[i];
            this.Data.News.push({
                ID: item.id,
                Data: item.questondate,
                Title: item.question,
                Author: item.questioner,

                Data2: item.answerdate,
                Author2: item.answerer,
                Content: item.answer,
                Symbol: item.symbol
            });

            if (i > 0) ++info.End;
        }

        this.CacheCount = this.Data.News.length;
        this.InvokeCallback(info);
    }

    this.RecvError = function (request) {
        console.log("[StockInteract::RecvError] ", request)
    }

    this.InvokeCallback = function (info) {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(info, this);
    }

    //下拉获取下一页
    this.GetNextPage = function () {
        var self = this;
        // 取最近1年的数据
        // var date = new Date();
        // var startDate = (date.getFullYear() - 1) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
        // var endDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

        var start = this.Data.News.length;
        var end = start + this.PageSize;

        if (typeof(this.Symbol) == "string") {
            var data = {
                "symbol": [this.Symbol],
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": this.SearchKey,
                "calccount": 1, //返回一共的数据个数
                "start": start,
                "end": end,
                "userid": this.userid
            }
        } else {
            var data = {
                "symbol": this.Symbol,
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": this.SearchKey,
                "calccount": 1, //返回一共的数据个数
                "start": start,
                "end": end,
                "userid": this.userid,
                "aggregate": this.Aggregate
            }
        }

        wx.request({
            url: this.ApiUrl,
            data: data,
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    //模块搜索
    this.ModuleSearch = function (search) {
        var self = this;

        //清空数据
        this.Data = {
            News: [],
            Aggregate: []
        };
        this.Count = 0;
        this.CacheCount = 0;
        // // 取最近1年的数据
        // var date = new Date();
        // var startDate = (date.getFullYear()) * 10000 + (date.getMonth() - 2) * 100 + date.getDate();
        // var endDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

        if (this.Symbol == null) {
            var data = {
                "symbol": [],
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": search,
                "calccount": 1, //返回一共的数据个数
                "start": 0,
                "end": this.PageSize,
                "userid": this.userid,
                "aggregate": this.Aggregate
            }
        } else {
            var data = {
                "symbol": [this.Symbol],
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": search,
                "calccount": 1, //返回一共的数据个数
                "start": 0,
                "end": this.PageSize,
                "userid": this.userid
            }
        }

        wx.request({
            url: this.ApiUrl,
            data: data,
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    //模块搜索获取下一页
    this.ModuleGetNextPage = function (search) {
        var self = this;
        // 取最近1年的数据
        // var date = new Date();
        // var startDate = (date.getFullYear()) * 10000 + (date.getMonth() - 2) * 100 + date.getDate();
        // var endDate = date.getFullYear() * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

        var start = this.Data.News.length;
        var end = start + this.PageSize;

        if (this.Symbol == null) {
            var data = {
                "symbol": [],
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": search,
                "calccount": 1, //返回一共的数据个数
                "start": start,
                "end": end,
                "userid": this.userid
            }
        } else {
            var data = {
                "symbol": [this.Symbol],
                "filed": ["symbol", "questioner", "question", "questondate", "answerer", "answer", "answerdate", "id"],
                // "querydate": {"StartDate": startDate, "EndDate": endDate},
                "querydate": {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                "search": search,
                "calccount": 1, //返回一共的数据个数
                "start": start,
                "end": end,
                "userid": this.userid
            }
        }

        wx.request({
            url: this.ApiUrl,
            data: data,
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }
}


//新闻详情
function StockNewsContent(id) {
    this.newMethod = INewsContent; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + "/API/NewsStockDetail2"

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "userid": null,
                "id": this.ID,
                "symbol": this.Symbol,
                "filed": ["name", "symbol", "releasedate", "title", "id", "content", "link", "source"]
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log("[StockNewsContent::RecvData] ",recvData);
        var data = recvData.data;
        var detail = data.detail;
        if (detail == null) return this.InvokeCallback();

        //把API数据格式 转换成自己的对外的统一格式,
        this.Data = {};
        this.Data.Date = detail.releasedate;
        this.Data.Content = detail.content;
        this.Data.Title = detail.title;
        this.Data.Source = detail.source;

        //新闻关联的其他股票列表
        var relation = new Array();
        for (var i in data.stocklist) {
            var item = data.stocklist[i];
            relation.push({
                Symbol: item.symbol,
                Name: item.name
            });
        }
        this.Data.Relation = relation;

        this.Name = detail.name;

        this.InvokeCallback();
    }

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    }

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    }
}

//调研详情
function marketSearchInfo(id) {
    this.newMethod = INewsContent; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + "/API/InvestorRelationsDetail"

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "userid": null,
                "id": this.ID,
              "filed": ["name", "symbol", "releasedate", "id", "content", "level", "researchdate", "type","researcher"]
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                if (data.data.code != 0) {
                    wx.showModal({
                        title: '提示',
                        content: data.data.message,
                    })
                    return;
                }
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log("[StockNewsContent::RecvData] ",recvData);
        var data = recvData.data;
        var detail = data.list[0];
        if (detail == null) return this.InvokeCallback();
        //把API数据格式 转换成自己的对外的统一格式,
        this.Data = {};
        this.Data.Releasedate = detail.releasedate;
        this.Data.Researchdate = detail.researchdate;
        this.Data.Content = detail.content;
        this.Data.Name = detail.name;
        this.Data.Symbol = detail.symbol;
        this.Data.Type = detail.type;
        this.Data.Level = detail.level;
        this.Data.Id = detail.id;
        this.Data.Researcher = detail.researcher;

        this.InvokeCallback();
    }

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    }

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    }
}


//负面新闻
function StockNegative(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsNegativeList';

    //获取第1页
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        let arySymbol = this.GetQuerySymbol();
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": arySymbol,
                "filed": ["symbol", "name", "releasedate", "title", "id", "source"],
                "querydate": {
                    "startDate": this.StartDate,
                    "endDate": this.EndDate
                },
                "search": null,
                "calccount": 1, //返回一共的数据个数
                "start": 0,
                "end": this.PageSize,
                "userid": null
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        var data = recvData.data;
        if (this.Count <= 0)
            this.Count = data.count; //一共的新闻个数

        var info = {};
        info.Start = this.Data.length;
        info.End = this.Data.length;

        for (var i in data.list) {
            var item = data.list[i];
            this.Data.push({
                Name: item.name,
                Symbol: item.symbol,
                ID: item.id,
                Data: item.releasedate,
                Title: item.title,
                Source: item.source
            });

            if (i > 0) ++info.End;
        }

        this.CacheCount = this.Data.length;
        this.InvokeCallback(info);
    }

    this.RecvError = function (request) {
        console.log("[StockNegative::RecvError] ", request)
    }

    this.InvokeCallback = function (info) {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(info, this);
    }

    //下拉获取下一页
    this.GetNextPage = function () {
        var self = this;

        let arySymbol = this.GetQuerySymbol();
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        var start = this.Data.length;
        var end = start + this.PageSize;
        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": arySymbol,
                "filed": ["symbol", "name", "releasedate", "title", "id", "source"],
                "querydate": {
                    "startDate": startDate,
                    "endDate": endDate
                },
                "search": null,
                "calccount": 0, //下页页不需要统计个数了
                "start": start,
                "end": end,
                "userid": null
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

}

//公告类型
var STOCK_REPORT_TYPE_ID = {
    REPORT_GOOD: '利好公告',
    REPORT_BAD: '利空公告'
};

//公告
function StockReport(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/ReportListByES'; //走公司内网查询接口
    this.ReportType = []; //公告类型数组

    //获取第1页
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        let arySymbol = this.GetQuerySymbol();
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        let reportType = null;
        if (this.ReportType && this.ReportType.length > 0) reportType = this.ReportType[0]; //暂时只支持1个类型

        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": arySymbol,
                "filed": ["symbol", "name", "releasedate", "title", "id", "source"],
                "querydate": {
                    "startDate": this.StartDate,
                    "endDate": this.EndDate
                },
                "search": null,
                "calccount": 1, //返回一共的数据个数
                "start": 0,
                "end": this.PageSize,
                "userid": null,
                'type': reportType
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        var data = recvData.data;
        if (this.Count <= 0)
            this.Count = data.count; //一共的新闻个数

        var info = {};
        info.Start = this.Data.length;
        info.End = this.Data.length;

        for (let i in data.report) {
            let item = data.report[i];
            this.Data.push({
                Name: item.name,
                Symbol: item.symbol,
                ID: item.id,
                Data: item.releasedate.substring(0, 4) + "-" + item.releasedate.substring(4, 6) + "-" + item.releasedate.substring(6, 8),
                Title: item.title,
                Source: item.source,
                Type: item.type,
                Showurl: item.showurl
            });

            if (i > 0) ++info.End;
        }

        this.CacheCount = this.Data.length;
        // this.InvokeCallback(info);
        this.InvokeCallback(this.Data);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.InvokeCallback = function (info) {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(info, this);
    }

    //下拉获取下一页
    this.GetNextPage = function () {
        var self = this;
        let arySymbol = this.GetQuerySymbol();
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        var start = this.Data.length;
        var end = start + this.PageSize;
        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": arySymbol,
                "filed": ["symbol", "name", "releasedate", "title", "id", "source"],
                "querydate": {
                    "startDate": self.StartDate,
                    "endDate": self.EndDate
                },
                "search": null,
                "calccount": 0, //下页页不需要统计个数了
                "start": start,
                "end": end,
                "userid": null
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

}

//新的负面新闻接口 NewsAllByES
function newStockReport(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/api/NewsAllByES_V2'; //走公司内网查询接口
    this.ReportType = []; //公告类型数组
    this.PageIndex = 1;
    this.PageSize = 10;
    this.SearchKey = "";
    this.Count;

    //获取第1页
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        let arySymbol = this.GetQuerySymbol();
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        let reportType = null;
        if (this.ReportType && this.ReportType.length > 0) reportType = this.ReportType[0]; //暂时只支持1个类型

        wx.request({
            url: this.ApiUrl,
            data: {
                "PageInfo": {
                    "PageIndex": 1,
                    "PageSize": this.PageSize
                },
                "Filter": {
                    "QueryDate": {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    "SymbolList": arySymbol,
                    "TypeList": this.ReportType, //["利空公告"],
                    "NegativeType": 10,
                    "SearchKey": this.SearchKey
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log("recvData", recvData);
        var data = recvData.data;
        if (this.Count <= 0)
            this.Count = data.count; //一共的新闻个数

        var info = {};
        info.Start = this.Data.length;
        info.End = this.Data.length;

        for (let i in data.list) {
            let item = data.list[i];
            this.Data.push({
                Name: item.stocklist[0].name,
                Symbol: item.stocklist[0].symbol,
                ID: item.id,
                Data: item.date,
                Title: item.title,
                Source: item.source,
                Type: item.type,
                Showurl: item.showurl,
                Negativetype: item.negativetype
            });

            if (i > 0) ++info.End;
        }

        this.CacheCount = this.Data.length;
        // this.InvokeCallback(info);
        this.InvokeCallback(this.Data);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.InvokeCallback = function (info) {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(info, this);
    }

    //下拉获取下一页
    this.GetNextPage = function () {

        console.log("this.PageIndex * this.PageSize", this.PageIndex * this.PageSize, this.Count)
        if (this.PageIndex * this.PageSize < this.Count) {
            this.PageIndex++;
            var self = this;
            let arySymbol = this.GetQuerySymbol();
            if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

            wx.request({
                url: this.ApiUrl,
                data: {
                    "PageInfo": {
                        "PageIndex": this.PageIndex,
                        "PageSize": this.PageSize
                    },
                    "Filter": {
                        "QueryDate": {
                            "StartDate": this.StartDate,
                            "EndDate": this.EndDate
                        },
                        "SymbolList": arySymbol,
                        "TypeList": this.ReportType, //["利空公告"],
                        "NegativeType": 10,
                        "SearchKey": this.SearchKey
                    }
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }

    }
}

//新的负面新闻详情 接口NewsNegativeDetail
function stockReportInfo(id) {
    this.newMethod = INewsBase; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsNegativeDetail';

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "id": this.ID,
                "filed": ["name", "symbol", "releasedate", "title", "id", "content", "source"]
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log("stockNew recvData",recvData);
        var data = recvData.data;
        var detail = data.list;
        if (detail.length == 0) return this.InvokeCallback();

        //把API数据格式 转换成自己的对外的统一格式,
        this.Data = {};
        this.Data.Date = detail[0].releasedate;
        this.Data.Content = clcontent(detail[0].content);
        this.Data.Title = detail[0].title;
        this.Data.Source = detail[0].source;
        this.Data.Symbol = detail[0].symbol;
        this.Data.Name = detail[0].name;

        //新闻关联的其他股票列表
        // var relation = new Array();
        // for (var i in data.stocklist) {
        //   var item = data.stocklist[i];
        //   relation.push({ Symbol: item.symbol, Name: item.name });
        // }
        // this.Data.Relation = relation;

        // this.Name = detail.name;

        this.InvokeCallback();
    }

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    }

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    }

}

//处理新的负面新闻详情 内容的格式
function clcontent(content) {
    console.log("处理新的负面新闻详情 内容的格式");
    // console.log(content);
    var reg = new RegExp("\r\n\r\n", "g");
    var regs = new RegExp("\r\n", "g");
    var regs1 = new RegExp("<br/><br/>", "g");
    var regs2 = new RegExp("\n", "g");
    var regs3 = new RegExp("    ", "g");
    var regs4 = new RegExp("\r", "g");
    content = content.replace(reg, '\r\n').replace(regs, '<br/>').replace(regs2, '<br/>').replace(regs1, '<br/>').replace(regs4, '<br/>').replace(regs3, ''); //&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
    var contentArr = content.split("<br/>");

    for (var i = contentArr.length - 1; i > 0; i--) {
        if (contentArr[i] == '') {
            contentArr.splice(i, 1);
        }
    }

    return contentArr;
}
// 处理官网新闻详情数据格式
function formatContent(content) {
    var reg = new RegExp("\n", "g");
    content = content.replace(reg, '<br/>');
    var contentArr = content.split("<br/>");
    for (let i = contentArr.length - 1; i > 0; i--) {
        if (contentArr[i] == '') {
            contentArr.splice(i, 1);
        }
    }
    return contentArr.map(e => e.trim());
}

//官网新闻
function OfficialWebsiteNews(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsCompanyList';
    this.PageIndex = 0;
    this.SearchKey = "";
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": this.Symbol,
                "filed": ["symbol", "name", "releasedate", "title", "id", "source", "link", "content"],
                "querydate": {
                    "startDate": this.StartDate,
                    "endDate": this.EndDate
                },
                "calccount": 1,
                "start": 0,
                "search": this.SearchKey,
                "aggregate": this.Aggregate
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        let data = recvData.data, formatData = [];
        if (data.list && data.list.length > 0) {
            formatData = data.list.map(e => Object.assign(e, {
                releasedate: e.releasedate.substring(0, 16)
            }));
        } else {
            formatData = [];
        }
        if (this.Count <= 0) this.Count = data.count; //一共的新闻个数
        this.Data = this.Data.concat(formatData);
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        if (this.PageSize * this.PageIndex <= this.Count) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    "symbol": this.Symbol,
                    "filed": ["symbol", "name", "releasedate", "title", "id", "source", "link", "content"],
                    "querydate": {
                        "startDate": this.StartDate,
                        "endDate": this.EndDate
                    },
                    "calccount": 1,
                    "search": this.SearchKey,
                    "aggregate": this.Aggregate,
                    "start": this.PageSize * this.PageIndex,
                    "end": this.PageSize * (this.PageIndex + 1),
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

//高管新闻
function SeniorExecutiveNews(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsTopManagersList';
    this.PageIndex = 0;
    this.SearchKey = "";
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                "symbol": this.Symbol,
                "filed": ["symbol", "name", "releasedate", "title", "id", "source"],
                "querydate": {
                    "startDate": this.StartDate,
                    "endDate": this.EndDate
                },
                "calccount": 1,
                "start": 0,
                "search": this.SearchKey,
                "aggregate": this.Aggregate
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        let data = recvData.data, formatData = [];
        if (data.list && data.list.length > 0) {
            formatData = data.list.map(e => Object.assign(e, {
                releasedate: e.releasedate.substring(0, 16)
            }));
        } else {
            formatData = [];
        }
        if (this.Count <= 0) this.Count = data.count; //一共的新闻个数
        this.Data = this.Data.concat(formatData);
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        if (this.PageSize * this.PageIndex <= this.Count) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    "symbol": this.Symbol,
                    "filed": ["symbol", "name", "releasedate", "title", "id", "source", "link", "content"],
                    "querydate": {
                        "startDate": this.StartDate,
                        "endDate": this.EndDate
                    },
                    "calccount": 1,
                    "search": this.SearchKey,
                    "aggregate": this.Aggregate,
                    "start": this.PageSize * this.PageIndex,
                    "end": this.PageSize * (this.PageIndex + 1),
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

//大V观点新闻
function ViewpointsByVIP(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsBigVList';
    this.PageIndex = 0;
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                "PageInfo": {
                    "PageIndex": this.PageIndex,
                    "PageSize": this.PageSize
                },
                "Filter": {
                    "QueryDate": {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    "CategoryList": []
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        let data = recvData.data, formatData = [];
        if (data.list && data.list.length > 0) {
            formatData = data.list.map(e => Object.assign(e, {
                date: e.date.substring(0, 16)
            }));
        } else {
            formatData = [];
        }
        if (this.Count <= 0) this.Count = data.count; //一共的新闻个数
        this.Data = this.Data.concat(formatData);
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        if (this.PageSize * this.PageIndex <= this.Count) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    "PageInfo": {
                        "PageIndex": this.PageIndex,
                        "PageSize": this.PageSize
                    },
                    "Filter": {
                        "QueryDate": {
                            "StartDate": this.StartDate,
                            "EndDate": this.EndDate
                        },
                        "CategoryList": []
                    }
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

//高管风险新闻概览
function TopManagersRiskNews(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/TopManagersRiskType';
    this.PageIndex = 0;
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                "Filter": {
                    "QueryDate": {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    "SymbolList": this.Symbol
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        let data = recvData.data;
        this.Data = data.list || [];
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        if (this.PageSize * this.PageIndex <= this.Count) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    "PageInfo": {
                        "PageIndex": this.PageIndex,
                        "PageSize": this.PageSize
                    },
                    "Filter": {
                        "QueryDate": {
                            "StartDate": this.StartDate,
                            "EndDate": this.EndDate
                        },
                        "SymbolList": this.Symbol
                    }
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

//官网新闻详情
function OfficialWebsiteInfo(id) {
    this.newMethod = INewsContent; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + "/API/NewsCompanyDetail";

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "userid": null,
                "id": this.ID,
                "filed": ["name", "symbol", "releasedate", "title", "id", "content", "link"]
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    };

    this.RecvData = function (recvData) {
        console.log("[StockNewsContent::RecvData] ", recvData);
        var data = recvData.data, detail = [];
        if (data.list && data.list.length > 0) {
            detail = data.list.map(e => Object.assign(e, {
                releasedate: e.releasedate.substring(0, 16),
                content: formatContent(e.content)
            }));
        } else {
            detail = [];
        }
        if (detail.length === 0) return this.InvokeCallback();
        this.Data = detail[0];
        console.log("this.Data", this.Data);
        this.InvokeCallback();
    };

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    };

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    };
}

//高管新闻详情
function SeniorExecutiveInfo(id) {
    this.newMethod = INewsContent; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + "/API/NewsTopManagersDetail";

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "userid": null,
                "id": this.ID,
                "filed": ["name", "symbol", "releasedate", "title", "id", "content", "link"]
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    };

    this.RecvData = function (recvData) {
        console.log("[StockNewsContent::RecvData] ", recvData);
        var data = recvData.data, detail = [];
        if (data.list && data.list.length > 0) {
            detail = data.list.map(e => Object.assign(e, {
                releasedate: e.releasedate.substring(0, 16),
                content: formatContent(e.content)
            }));
        } else {
            detail = [];
        }
        if (detail.length === 0) return this.InvokeCallback();
        this.Data = detail[0];
        this.InvokeCallback();
    };

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    };

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    };
}

//大V观点新闻详情
function ViewpointsByVIPInfo(id) {
    this.newMethod = INewsContent; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + "/API/NewsBigVDetail";

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "NewsId": this.ID,
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    };

    this.RecvData = function (recvData) {
        console.log("[StockNewsContent::RecvData] ", recvData);
        var data = recvData.data;
        if (data.code !== 0) return this.InvokeCallback();
        this.Data = {
            date: data.date.substring(0, 16),
            title: data.title,
            source: data.source,
            content: formatContent(data.content)
        };
        console.log("this.Data", this.Data);
        this.InvokeCallback();
    };

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    };

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    };
}
//高管风险新闻详情
function TopManagersRiskNewsInfo(id) {
    this.newMethod = INewsContent; //派生
    this.newMethod(id);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + "/API/NewsTopManagersRiskDetail";

    this.GetContent = function () {
        var self = this;
        //清空数据
        this.Name = null;
        this.Error = null;
        this.Data = null;

        wx.request({
            url: this.ApiUrl,
            data: {
                "NewsId": this.ID,
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    };

    this.RecvData = function (recvData) {
        console.log("[StockNewsContent::RecvData] ", recvData);
        var data = recvData.data;
        if (data.code !== 0) return this.InvokeCallback();
        this.Data = {
            date: data.date.substring(0, 16),
            title: data.title,
            name: data.name,
            source: data.source,
            symbol: data.symbol,
            content: formatContent(data.content)
        };
        console.log("this.Data", this.Data);
        this.InvokeCallback();
    };

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    };

    this.InvokeCallback = function () {
        if (typeof(this.Callback) != 'function') return;

        this.Callback(this);
    };
}


//重组失败
function FailureRegroupNews(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/ReportListByES_V2';
    this.PageIndex = 1;
    this.SearchKey = "";
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                PageInfo: {
                    PageIndex: this.PageIndex,
                    PageSize: 10
                },
                Filter: {
                    SymbolList: this.Symbol,
                    QueryDate: {
                        StartDate: this.StartDate,
                        EndDate: this.EndDate
                    },
                    Search: {
                        OrList: [
                            "终止重组",
                            "重组终止",
                            "重组失败",
                            "重组取消"
                        ],
                        AndList: this.SearchKey ? [this.SearchKey] : []
                    }
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        let data = recvData.data, formatData = [];
        if (data.list && data.list.length > 0) {
            formatData = data.list.map(e => Object.assign(e, {
                showurl: e.showurl ? e.showurl.replace('http://rpt.zealink.com', 'https://reporth5.zealink.com') : ''
            }));
        } else {
            formatData = [];
        }
        if (this.Count <= 0) this.Count = data.count; //一共的新闻个数
        this.Data = this.Data.concat(formatData);
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        var pages = this.Count / this.PageSize,
            totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
        if (this.PageIndex <= totalPages) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    PageInfo: {
                        PageIndex: this.PageIndex,
                        PageSize: 10
                    },
                    Filter: {
                        SymbolList: this.Symbol,
                        QueryDate: {
                            StartDate: this.StartDate,
                            EndDate: this.EndDate
                        },
                        Search: {
                            OrList: [
                                "终止重组",
                                "重组终止",
                                "重组失败",
                                "重组取消"
                            ],
                            AndList: this.SearchKey ? [this.SearchKey] : []
                        }
                    }
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

//请求动态数据
function GetNews() {
    // this.ApiUrl = g_JSNewsResource.Domain + '/api/NewsFlashListByES';

    var date = new Date();
    var today = (date.getFullYear()) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
    date.setDate(date.getDate() - 5);
    var fiveDaysAgo = (date.getFullYear()) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();

    this.Count; //一共的新闻个数
    this.PageSize = 20; //每页的新闻个数
    this.CacheCount; //已下载的新闻个数
    this.StartDate = "";
    this.EndDate = "";
    this.SymbolList = [];
    this.PageIndex = 1;
    this.TypeList = [];
    this.ElseTypeList = [];
    this.SearchKey = "";
    this.NowData = 0; //当前展示第几页的数据
    this.IsAutoUpdate = true; //定时刷新
    this.ShowValue = 1,


        this.Stop = function () {
            // console.log("[JSStock::Stop] stop update.")
            this.IsAutoUpdate = false;
            if (this.Timeout) clearTimeout(this.Timeout); //清空定时器
        }

    this.Start = function () {
        // console.log("[JSStock::Start] start auto update.");
        this.IsAutoUpdate = true;
        this.GetFirstPage();
    }

    //请求数据
    this.GetFirstPage = function () {
        this.NowData = 0;

        var self = this;
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;
        // this.ShowValue = 1,
        wx.request({
            url: g_JSNewsResource.Domain + '/api/NewsFlashListByES_V2',
            // url: 'https://apitest.zealink.com/api/NewsFlashListByES_V2',
            data: {
                "ShowValue": self.ShowValue,
                "PageInfo": {
                    "PageIndex": self.PageIndex,
                    "PageSize": self.PageSize
                },
                "Filter": {
                    "QueryDate": {
                        "StartDate": self.StartDate,
                        "EndDate": self.EndDate
                    },
                    "SymbolList": self.SymbolList,
                    "Typelist": self.TypeList,
                    "SearchKey": self.SearchKey
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                // if (data.data.code != 0) {
                //   wx.showModal({
                //     title: '提示',
                //     content: data.data.message,
                //   })
                //   // 隐藏加载框
                //   wx.hideLoading();
                //   return;
                // }
                self.RecvData(data);
            },
            fail: function (request) {
                self.RecvError(request);
            }
        })
    },

        this.RecvData = function (recvData) {
            // console.log(recvData.data,"快讯recvData")
            if (recvData.data.isshow == false) return this.InvokeCallback();
            if (recvData.data.code !== 0) return this.InvokeCallback();
            if (recvData.data.list.length <= 0) return this.InvokeCallback();

            for (var i in recvData.data.list) {
                var item = recvData.data.list[i];
                var info = {};
                info.Date = item.date;
                info.Time = item.time;
                info.Sourceid = item.sourceid;
                info.Stock = item.stock;
                info.Summary = item.summary;
                info.Title = item.title;
                info.Type = item.type;
                info.Image = item.image;
                this.Data.push(info);
            }

            if (this.Count <= 0) this.Count = recvData.data.count; //一共的新闻个数

            this.InvokeCallback();

            var self = this;
            if (this.IsAutoUpdate) {
                if (this.Timeout) clearTimeout(this.Timeout); //清空定时器
                this.Timeout = setTimeout(function () {
                    self.GetFirstPage();
                }, 5000);
            }
        },

        this.RecvError = function (reqeust) {
            this.Error = "请求失败";
            this.InvokeCallback();
        },

        this.InvokeCallback = function () {
            if (typeof(this.Callback) != 'function') return;

            this.Callback(this);
        },
        this.GetNextPage = function () {
            this.PageIndex++;
            var self = this;
            this.NowData = 1;
            var pages = this.Count / this.PageSize,
                totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
            if (this.PageIndex <= totalPages) {
                wx.request({
                    url: g_JSNewsResource.Domain + '/api/NewsFlashListByES_V2',
                    // url: 'https://apitest.zealink.com/api/NewsFlashListByES_V2',
                    data: {
                        "ShowValue": self.ShowValue,
                        "PageInfo": {
                            "PageIndex": self.PageIndex,
                            "PageSize": self.PageSize
                        },
                        "Filter": {
                            "QueryDate": {
                                "StartDate": self.StartDate,
                                "EndDate": self.EndDate
                            },
                            "SymbolList": self.SymbolList,
                            "Typelist": self.TypeList,
                            "SearchKey": self.SearchKey
                        }
                    },
                    method: "POST",
                    dataType: "json",
                    success: function (data) {
                        self.RecvData(data);
                        wx.hideLoading();
                    },
                    error: function (request) {
                        self.RecvError(request);
                    }
                })
            } else {
                wx.showToast({
                    title: "数据已全部加载",
                    icon: 'success',
                    duration: 1000
                })
            }
        }
}
GetNews.init = function () {
    var news = new GetNews();
    return news;
}


//请求新闻联播列表
function GetCCTVList() {
    this.Count; //一共的新闻个数
    this.CacheCount; //已下载的新闻个数
    this.SearchKey = ""; //搜索关键词
    this.PageIndex = 1; //请求第几页
    this.PageSize = 10; //每页数量
    var date = new Date();
    var nowDate = (date.getFullYear()) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
    this.Date = nowDate; //查询日期
    this.Url = g_JSNewsResource.JavaDomain + "/news/list";


    //请求第一页数据
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;

        wx.request({
            url: self.Url,
            data: {
                "key": self.SearchKey,
                "page": 1,
                "rows": self.PageSize,
                "sdate": self.Date,
                "edate": self.Date
            },
            method: "POST",
            dataType: "json",
            success: function (res) {
                if (res.data.code != 0) {
                    wx.showModal({
                        title: '提示',
                        content: res.data.msg,
                    })
                    return;
                }
                self.RecvData(res);
            },
            fail: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log(recvData.data,"新闻列表")
        var data = recvData.data;
        if (data.data.count) this.Count = data.data.count;
        if (data.data.newslist.length != 0 && data.data.newslist.length != undefined) {
            for (var i in data.data.newslist) {
                var item = data.data.newslist[i];
                var info = {};
                info.Index = item.f001N;
                info.Title = item.f002V;
                info.Content = item.f003V;
                info.Date = item.tdate;

                this.Data.push(info);
            }
        }

        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (reqeust) {
        console.log(reqeust, "error:")
    }

    this.GetNextPage = function () {
        this.PageIndex++;

        var self = this;

        var pages = this.Count / this.PageSize,
            totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
        if (this.PageIndex <= totalPages) {
            wx.request({
                url: self.Url,
                data: {
                    "key": self.SearchKey,
                    "page": self.PageIndex,
                    "rows": self.PageSize,
                    "sdate": self.Date,
                    "edate": self.Date
                },
                method: "POST",
                dataType: "json",
                success: function (res) {
                    if (res.data.code != 0) {
                        wx.showModal({
                            title: '提示',
                            content: res.data.msg,
                        })
                        return;
                    }
                    self.RecvData(res);
                },
                fail: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }

}

GetCCTVList.init = function () {
    var news = new GetCCTVList();
    return news;
}

//请求 CCTV 热点
function GetCCTVHot() {
    var date = new Date();
    var newDate = (date.getFullYear()) * 10000 + (date.getMonth() + 1) * 100 + date.getDate();
    this.Date = newDate.toString(); //查询日期

    this.requestData = function () {
        //清空数据
        this.Data = [];
        var self = this;

        wx.request({
            url: g_JSNewsResource.JavaDomain + "/cctv/getHotWords",
            data: {
                "date": self.Date
            },
            method: "POST",
            dataType: "json",
            success: function (res) {
                if (res.data.code != 0) {
                    wx.showModal({
                        title: '提示',
                        content: res.data.msg,
                    })
                    return;
                }
                self.RecvData(res);
            },
            fail: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log(recvData.data, "CCTV热点:")
        var data = recvData.data;
        var info = {};
        if (data.data.PreSevenHotWords) {
            var SevenHotWords = [];
            for (var i in data.data.PreSevenHotWords) {
                var item = data.data.PreSevenHotWords[i];
                var SevenWord = {};
                SevenWord.Name = item.f001v;
                SevenWord.Count = item.f002n;
                SevenHotWords.push(SevenWord);
            }
            info.PreSevenHotWords = SevenHotWords;
        }
        if (data.data.PreTenHotWords) {
            var TenHotWords = [];
            for (var i in data.data.PreTenHotWords) {
                var item = data.data.PreTenHotWords[i];
                var TenWord = {};
                TenWord.Name = item.f001v;
                TenWord.Count = item.f002n;
                TenWord.Date = item.tdate;
                TenHotWords.push(TenWord);
            }
            info.PreTenHotWords = TenHotWords;
        }
        this.Data = info;

        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (reqeust) {
        console.log(reqeust, "error:")
    }
}

GetCCTVHot.init = function () {
    var news = new GetCCTVHot();
    return news;
}

//获取最近三个月的数量统计和新闻列表
function GetThreeMonthNews() {
    this.Count; //一共的新闻个数
    this.CacheCount; //已下载的新闻个数
    this.SearchKey; //搜索关键词
    this.PageIndex = 1; //请求第几页
    this.PageSize = 10; //每页数量
    this.Url = g_JSNewsResource.JavaDomain + "/news/hotNews";
    this.GetNum = 0; //请求第几页的数据

    var startData = JSCommonDate.startDate(null, -3, "m");
    var today = JSCommonDate.endTime();

    this.startData = startData;
    this.endData = today;

    //请求第一页数据
    this.GetFirstPage = function () {
        this.GetNum = 0;
        //清空数据
        this.Data = {
            HotwordList: [],
            NewsList: []
        };
        this.Count = 0;
        this.CacheCount = 0;
        var self = this;

        wx.request({
            url: self.Url,
            data: {
                "key": self.SearchKey,
                "page": 1,
                "rows": self.PageSize,
                "sdate": this.startData,
                "edate": this.endData
            },
            method: "POST",
            dataType: "json",
            success: function (res) {
                if (res.data.code != 0) {
                    wx.showModal({
                        title: '提示',
                        content: res.data.msg,
                    })
                    return;
                }
                self.RecvData(res);
            },
            fail: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log(recvData.data,"recvData最近三个月")
        var data = recvData.data;
        if (data.data.count) this.Count = data.data.count;
        if (this.GetNum == 0) {
            if (data.data.hotwordlist) this.Data.HotwordList = data.data.hotwordlist;
        }

        if (data.data.newslist) {
            for (var i in data.data.newslist) {
                var item = data.data.newslist[i];
                var info = {};
                info.Index = item.f001N;
                info.Title = item.f002V;
                info.Content = item.f003V;
                info.Date = item.tdate;

                this.Data.NewsList.push(info);
            }
        }

        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (reqeust) {
        console.log(reqeust, "error:")
    }

    this.GetNextPage = function () {
        this.GetNum = 1;
        this.PageIndex++;

        var pages = this.Count / this.PageSize,
            totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
        var self = this;
        if (this.PageIndex <= totalPages) {
            wx.request({
                url: self.Url,
                data: {
                    "key": self.SearchKey,
                    "page": self.PageIndex,
                    "rows": self.PageSize,
                    "sdate": this.startData,
                    "edate": this.endData
                },
                method: "POST",
                dataType: "json",
                success: function (res) {
                    if (res.data.code != 0) {
                        wx.showModal({
                            title: '提示',
                            content: res.data.msg,
                        })
                        return;
                    }
                    self.RecvData(res);
                },
                fail: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }

}

GetThreeMonthNews.init = function () {
    var news = new GetThreeMonthNews();
    return news;
}

//请求  原创 数据
function GetOriginal() {
    this.Count; //一共的新闻个数
    this.CacheCount; //已下载的新闻个数
    this.PageIndex = 1; //请求第几页
    this.PageSize = 10; //每页数量
    var startData = JSCommonDate.startDate(null, -3, "m");
    var today = JSCommonDate.endTime();
    this.StartDate = startData; //查询开始日期
    this.EndDate = today;       //查询结束日期
    this.Url = g_JSNewsResource.Domain + '/api/NewsFlashOriginalList';

    //请求数据
    this.GetFirstPage = function () {
        this.NowData = 0;

        var self = this;
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;
        wx.request({
            url: this.Url,
            data: {
                "PageInfo": {
                    "PageIndex": 1,
                    "PageSize": self.PageSize
                },
                "Filter": {
                    "QueryDate": {
                        "StartDate": self.StartDate,
                        "EndDate": self.EndDate
                    }
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                if (data.data.code != 0) {
                    wx.showModal({
                        title: '提示',
                        content: data.data.message,
                    })
                    // 隐藏加载框
                    wx.hideLoading();
                    return;
                }
                self.RecvData(data);
            },
            fail: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log(recvData.data,"原创recvData")
        if (recvData.data.list.length <= 0) return this.InvokeCallback();
        if (this.Count <= 0) this.Count = recvData.data.count; //一共的新闻个数

        for (var i in recvData.data.list) {
            var item = recvData.data.list[i];
            var info = {};
            info.Date = item.date;
            info.ID = item.id;
            info.Summary = item.summary;
            info.Title = item.title;
            info.Image = item.image;
            info.URL = item.url;
            this.Data.push(info);
        }

        this.InvokeCallback();
    }

    this.RecvError = function (reqeust) {
        this.Error = "请求失败";
        this.InvokeCallback();
    }

    this.InvokeCallback = function () {
        if (typeof (this.Callback) != 'function') return;

        this.Callback(this);
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        var pages = this.Count / this.PageSize,
            totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);

        if (this.PageIndex <= totalPages) {
            wx.request({
                url: this.Url,
                data: {
                    "PageInfo": {
                        "PageIndex": self.PageIndex,
                        "PageSize": self.PageSize
                    },
                    "Filter": {
                        "QueryDate": {
                            "StartDate": self.StartDate,
                            "EndDate": self.EndDate
                        }
                    }
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    if (data.data.code != 0) {
                        wx.showModal({
                            title: '提示',
                            content: data.data.message,
                        })
                        // 隐藏加载框
                        wx.hideLoading();
                        return;
                    }
                    self.RecvData(data);
                },
                fail: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

GetOriginal.init = function () {
    var news = new GetOriginal();
    return news;
}

//条件选股
function PickStock(){
  this.Count; //总数
  this.CacheCount; //已下载的个数
  this.StartSize = 0; //请求开始数据
  this.PageSize = 20; //每页显示数据
  this.ID = "";
  this.Block = "";    //block
  this.Index = [];   //index
  this.Condition = [];  //condition
  this.Orderfield = "price"; 
  this.Order = -1;

  this.Url = g_JSNewsResource.Domain + '/api/conditionStock';

  this.getFirstData = function(){
    var self = this;
    //清空数据
    this.Data = {list:[],count:0};
    this.Count = 0;
    this.CacheCount = 0;

    var data;
    if (this.Index.length == 0 && this.Condition.length == 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": this.StartSize,
        "end": this.PageSize
      }
    } else if (this.Index.length > 0 && this.Condition.length == 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "index": this.Index,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": this.StartSize,
        "end": this.PageSize
      }
    } else if (this.Index.length == 0 && this.Condition.length > 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "condition": this.Condition,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": this.StartSize,
        "end": this.PageSize
      }
    } else if (this.Index.length > 0 && this.Condition.length > 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "index": this.Index,
        "condition": this.Condition,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": this.StartSize,
        "end": this.PageSize
      }
    }

    wx.request({
      url: this.Url,
      data: data,
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          // 隐藏加载框
          wx.hideLoading();
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"条件选股recvData")
    if (recvData.data.list == null) return this.InvokeCallback();
    if (this.Count <= 0) {
      this.Count = recvData.data.count; //一共个数
      this.Data.count = recvData.data.count;
    }

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.Name = item.name;
      info.Symbol = item.symbol;
      for (var j in item.field){
        var ite = item.field[j];
        if (ite.name == "price") {
          info.Price = ite.value;
        } else if (ite.name == "amount"){
          info.Amount = ite.value;
        } else if (ite.name == "increase"){
          info.Increase = ite.value;
        }
      }
      this.Data.list.push(info);
    }

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function(){
    var self = this;
    var start = this.Data.list.length;
    var end = start + this.PageSize;

    var data;
    if (this.Index.length == 0 && this.Condition.length == 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": start,
        "end": end
      }
    } else if (this.Index.length > 0 && this.Condition.length == 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "index": this.Index,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": start,
        "end": end
      }
    } else if (this.Index.length == 0 && this.Condition.length > 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "condition": this.Condition,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": start,
        "end": end
      }
    } else if (this.Index.length > 0 && this.Condition.length > 0) {
      data = {
        "userid": this.ID,
        "block": this.Block,
        "index": this.Index,
        "condition": this.Condition,
        "field": ["price","amount", "increase"],
        "orderfield": this.Orderfield,
        "order": this.Order,
        "start": start,
        "end": end
      }
    }

    if (this.Data.list.length < this.Count){
      wx.request({
        url: this.Url,
        data: data,
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    }else{
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }

  }
}

PickStock.init = function () {
  var news = new PickStock();
  return news;
}

//添加策略
function AddCondition(){
  this.ID = "";
  this.PolicyName = "",  //策略名称
  this.MarketList = "";    //市场集合
  this.IndexList = [];   //指标集合
  this.ConditionList = [];  //条件集合
  this.Url = g_JSNewsResource.Domain + '/api/SP_AddUserPolicy';

  this.requestData = function(){
    var data = {
      "UserId": this.ID,
      "PolicyName": this.PolicyName,
      "Market": this.MarketList,
      "ConditionList": this.ConditionList,
      "IndexList": this.IndexList
    }

    wx.request({
      url: this.Url,
      data: data,
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        // self.RecvData(data);
      },
      fail: function (request) {
        // self.RecvError(request);
      }
    })
  }
}

AddCondition.init = function () {
  var news = new AddCondition();
  return news;
}

function GetCondition(){
  this.Count; //总数
  this.CacheCount; //已下载的个数
  this.ID = "";
  this.PageIndex = 1;
  this.PageSize = 20;

  this.Url = g_JSNewsResource.Domain + '/api/SP_UserPolicyList';

  this.getFirstData = function(){
    var self = this;
    //清空数据
    this.Data = [];
    this.Count = 0;
    this.CacheCount = 0;

    wx.request({
      url: this.Url,
      data: {
        "UserId": this.ID,
        "PageInfo": {
          "PageIndex": 1,
          "PageSize": this.PageSize
        }
      },
      method: "POST",
      dataType: "json",
      success: function (data) {
        if (data.data.code != 0) {
          wx.showModal({
            title: '提示',
            content: data.data.message,
          })
          return;
        }
        self.RecvData(data);
      },
      fail: function (request) {
        self.RecvError(request);
      }
    })
  }

  this.RecvData = function (recvData) {
    // console.log(recvData.data,"获取自建策略recvData")
    if (recvData.data.list.length == 0) return this.InvokeCallback();
    if (this.Count <= 0) this.Count = recvData.data.count; //一共个数

    for (var i in recvData.data.list) {
      var item = recvData.data.list[i];
      var info = {};
      info.Guid = item.guid;
      info.Indexlist = item.indexlist;
      info.Market = item.market;
      info.Policyname = item.policyname;
      info.Conditionlist = item.conditionlist;

      this.Data.push(info);
    }
    

    this.InvokeCallback();
  }

  this.RecvError = function (reqeust) {
    this.Error = "请求失败";
    this.InvokeCallback();
  }

  this.InvokeCallback = function () {
    if (typeof (this.Callback) != 'function') return;

    this.Callback(this);
  }

  this.getNextData = function(){
    this.PageIndex++;
    var self = this;
    var pages = this.Count / this.PageSize,
      totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
    
    if (this.PageIndex <= totalPages){
      wx.request({
        url: this.Url,
        data: {
          "UserId": this.ID,
          "PageInfo": {
            "PageIndex": this.PageIndex,
            "PageSize": this.PageSize
          }
        },
        method: "POST",
        dataType: "json",
        success: function (data) {
          self.RecvData(data);
        },
        fail: function (request) {
          self.RecvError(request);
        }
      })
    }else{
      wx.showToast({
        title: "数据已全部加载",
        icon: 'success',
        duration: 1000
      })
    }
  }
}

GetCondition.init = function () {
  var news = new GetCondition();
  return news;
}


//高管风险类型新闻
function TopManagersRiskList(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    // this.ApiUrl = g_JSNewsResource.Domain + '/API/NewsTopManagersRiskList';
    this.ApiUrl = 'https://apitest.zealink.com/API/NewsTopManagersRiskList';
    this.PageIndex = 0;
    this.TypeList = [];
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                "PageInfo": {
                    "PageIndex": this.PageIndex,
                    "PageSize": this.PageSize
                },
                "Filter": {
                    "QueryDate": {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    "SymbolList": this.Symbol,
                    "TypeList": this.TypeList
                }
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        let data = recvData.data;
        if (this.Count <= 0) this.Count = data.count; //一共的新闻个数
        this.Data = this.Data.concat(data.list);
        console.log("this.Data上拉", this.Data);
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        if (this.PageSize * this.PageIndex <= this.Count) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    "PageInfo": {
                        "PageIndex": this.PageIndex,
                        "PageSize": this.PageSize
                    },
                    "Filter": {
                        "QueryDate": {
                            "StartDate": this.StartDate,
                            "EndDate": this.EndDate
                        },
                        "SymbolList": this.Symbol,
                        "TypeList": this.TypeList
                    }
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}

// 公告搜索
function SearchNotice(symbol) {
    this.newMethod = INewsBase; //派生
    this.newMethod(symbol);
    delete this.newMethod;

    this.ApiUrl = g_JSNewsResource.Domain + '/API/reportStockList'; // g_JSNewsResource.Domain + '/API/reportStockList' ;   'http://web7.umydata.com/API/reportStockList'
    this.PageIndex = 0;
    this.SearchKey = "";
    this.userid = "";
    this.Aggregate = [];
    this.GetFirstPage = function () {
        //清空数据
        this.Data = [];
        this.Count = 0;
        this.CacheCount = 0;

        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();
        wx.request({
            url: this.ApiUrl,
            data: {
                symbol: this.symbol,
                filed: ["name", "symbol", "releasedate", "title", "id"],
                querydate: {
                    "StartDate": this.StartDate,
                    "EndDate": this.EndDate
                },
                calccount: 1,
                start: this.PageSize * this.PageIndex,
                end: this.PageSize * (this.PageIndex + 1),
                search: this.SearchKey,
                userid: this.userid,
                aggregate: 10
            },
            method: "POST",
            dataType: "json",
            success: function (data) {
                self.RecvData(data);
            },
            error: function (request) {
                self.RecvError(request);
            }
        })
    }

    this.RecvData = function (recvData) {
        // console.log(recvData.data.list,"前");
        let data = recvData.data,
            aggregate = data.aggregate,
            formatData = data.report.map(e => Object.assign(e, {
                showurl: e.showurl ? e.showurl.replace('http://rpt.zealink.com', 'https://reporth5.zealink.com') : '',
                releasedate: e.releasedate.substring(0, 4) + "-" + e.releasedate.substring(4, 6) + "-" + e.releasedate.substring(6, 8)
            }));
        console.log(formatData, "后");
        if (this.Count <= 0) this.Count = data.count; //一共的新闻个数

        if (aggregate) {
            for (let i in aggregate) {
                this.Aggregate.push({
                    Name: aggregate[i].value2,
                    Symbol: aggregate[i].value,
                    Count: aggregate[i].count
                });
            }
        }

        this.Data = this.Data.concat(formatData);
        //console.log("this.Data上拉",this.Data);
        if (this.Callback) this.Callback(this);
    }

    this.RecvError = function (request) {
        console.log("[StockReport::RecvError] ", request)
    }

    this.GetNextPage = function () {
        this.PageIndex++;
        var self = this;
        if (!this.StartDate || !this.EndDate) this.SetDefautQueryDate();

        var pages = this.Count / this.PageSize,
            totalPages = (this.Count % this.PageSize) > 0 ? Math.floor(pages) + 1 : Math.floor(pages);
        if (this.PageIndex <= totalPages) {
            wx.request({
                url: this.ApiUrl,
                data: {
                    symbol: this.symbol,
                    filed: ["name", "symbol", "releasedate", "title", "id"],
                    querydate: {
                        "StartDate": this.StartDate,
                        "EndDate": this.EndDate
                    },
                    calccount: 1,
                    start: this.PageSize * this.PageIndex,
                    end: this.PageSize * (this.PageIndex + 1),
                    search: this.SearchKey,
                    userid: this.userid,
                    aggregate: 10
                },
                method: "POST",
                dataType: "json",
                success: function (data) {
                    self.RecvData(data);
                },
                error: function (request) {
                    self.RecvError(request);
                }
            })
        } else {
            wx.showToast({
                title: "数据已全部加载",
                icon: 'success',
                duration: 1000
            })
        }
    }
}


module.exports = {
    JSCommonNews: {
      JSNews: JSNews,
      StockNews: StockNews,
      StockInteract: StockInteract,
      NEWS_TYPE: NEWS_TYPE,
      STOCK_REPORT_TYPE_ID: STOCK_REPORT_TYPE_ID,
      GetNews: GetNews,
      GetCCTVList: GetCCTVList,
      GetCCTVHot: GetCCTVHot,
      GetThreeMonthNews: GetThreeMonthNews,
      GetOriginal: GetOriginal,
      PickStock: PickStock,
      AddCondition: AddCondition,
      GetCondition: GetCondition
    }
};
